{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Utils\n",
    "\n",
    "> This contains useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import gffpandas.gffpandas as gffpd\n",
    "import numpy as np\n",
    "\n",
    "from bokeh.plotting import figure\n",
    "from bokeh.models.tools import BoxZoomTool\n",
    "from bokeh.models import HoverTool, NumeralTickFormatter, LabelSet\n",
    "from bokeh.models.glyphs import Patches\n",
    "from bokeh.models import (\n",
    "    CustomJS,\n",
    "    Range1d,\n",
    "    ColumnDataSource,\n",
    ")\n",
    "\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "def create_genome_browser_plot(glyphSource, x_range, **kwargs):\n",
    "    plot_height = kwargs.get(\"plot_height\", 150)\n",
    "    label_angle = kwargs.get(\"label_angle\", 45)\n",
    "    text_font_size = kwargs.get(\"text_font_size\", \"10pt\")\n",
    "    output_backend = kwargs.get(\"output_backend\", \"webgl\")\n",
    "    \n",
    "    y_min, y_max = get_y_range()\n",
    "    p_annot = figure(\n",
    "        tools = \"xwheel_zoom,xpan,save\",\n",
    "        active_scroll = \"xwheel_zoom\",\n",
    "        height = plot_height,\n",
    "        x_range = x_range,\n",
    "        y_range = Range1d(y_min, y_max),\n",
    "        output_backend=output_backend,\n",
    "    )\n",
    "    # Add tool\n",
    "    p_annot.add_tools(BoxZoomTool(dimensions=\"width\"))\n",
    "\n",
    "    #p_annot.sizing_mode = \"stretch_both\"\n",
    "\n",
    "    # Format x axis values\n",
    "    p_annot.xaxis[0].formatter = NumeralTickFormatter(format=\"0,0\")\n",
    "    # Hide grid\n",
    "    p_annot.xgrid.visible = False\n",
    "    p_annot.ygrid.visible = False\n",
    "    # Hide axis\n",
    "    p_annot.yaxis.visible = False\n",
    "    glyph = p_annot.add_glyph(\n",
    "        glyphSource, Patches(xs=\"xs\", ys=\"ys\", fill_color=\"color\")\n",
    "    )\n",
    "    # gene labels in the annotation track\n",
    "    # This seems to be necessary to show the labels\n",
    "    p_annot.scatter(x=\"pos\", y=0, size=0, source=glyphSource)\n",
    "    labels = LabelSet(\n",
    "        x=\"pos\",\n",
    "        y=-0.9,\n",
    "        text=\"names\",\n",
    "        level=\"glyph\",\n",
    "        angle=label_angle,\n",
    "        text_font_size=text_font_size,\n",
    "        x_offset=-5,\n",
    "        y_offset=0,\n",
    "        source=glyphSource,\n",
    "        text_align='left',\n",
    "    )\n",
    "\n",
    "    p_annot.add_layout(labels)\n",
    "    p_annot.add_tools(\n",
    "        HoverTool(\n",
    "            renderers=[glyph],\n",
    "            tooltips=[(\"locus_tag\", \"@locus_tag\"), (\"gene\", \"@gene\"), (\"product\", \"@product\")],\n",
    "        )\n",
    "    )\n",
    "    return p_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_genome_annotations(gff_path: str, seq_id: str=None, bounds=None):\n",
    "    annotation = gffpd.read_gff3(gff_path)\n",
    "    annotation = annotation.df\n",
    "    if seq_id:\n",
    "        annotation = annotation.loc[(annotation.seq_id == seq_id)]\n",
    "        if len(annotation)==0:\n",
    "            warnings.warn(\"The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\")\n",
    "        \n",
    "        \n",
    "    if bounds:\n",
    "        annotation = annotation.loc[(annotation.start<bounds[1]) & (annotation.end>bounds[0])]\n",
    "\n",
    "    annotation.loc[:, \"left\"] = annotation[[\"start\"]].values\n",
    "    annotation.loc[:, \"right\"] = annotation[[\"end\"]].values\n",
    "    return annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from genomenotebook.js_callback_code import get_example_data_dir\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  left    right  \n",
       "0  ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...     1  4641652  \n",
       "1  ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...   190      255  \n",
       "2  ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...   190      255  \n",
       "3  ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...   337     2799  \n",
       "4  ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...   337     2799  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "annotation=get_genome_annotations(gff_path, \"U00096.3\")\n",
    "annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3778/2279574846.py:8: UserWarning: The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\n",
      "  warnings.warn(\"The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\")\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "annotation=get_genome_annotations(gff_path, \"U00097.3\") #mistake in seq_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_attribute(input_str:str, #attribute string to parse\n",
    "                      attr_name:str, #name of the attribute to extract\n",
    "                     ) -> dict:\n",
    "    \"\"\"Extracts the attribute called attr_name from the GFF attributes string\"\"\"\n",
    "    \n",
    "    pattern = f\"[{attr_name[0].lower()}{attr_name[0].upper()}]{attr_name[1:]}=(?P<{attr_name}>[^;]+)\"\n",
    "    match = re.search(pattern, input_str)\n",
    "    if match:\n",
    "        return match.groupdict()[attr_name]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dnaA'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "input_str = 'ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxref=NCBI_GP:ATV02827.1;Name=ATV02827.1;gbkey=CDS;gene=dnaA;locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "extract_attribute(input_str,\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: attributes, dtype: object)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = annotation[annotation.type.isin([\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"])    ]\n",
    "genes.attributes.apply(extract_attribute,attr_name='protein_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "input_str = 'locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "assert extract_attribute(input_str,\"gene\") == None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_genes_from_annotation(annotation):\n",
    "    \"\"\"Filters annotations to keep only features of type CDS, repeat_region, ncRNA, rRNA and tRNA.\n",
    "       Extracts gene, locus_tag and product attributes as columns.\n",
    "    \"\"\"\n",
    "    genes = annotation[\n",
    "        annotation.type.isin([\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"])\n",
    "    ].copy()\n",
    "\n",
    "    genes.loc[genes[\"strand\"] == \"+\", \"start\"] = genes.loc[\n",
    "        genes[\"strand\"] == \"+\", \"left\"\n",
    "    ].values\n",
    "\n",
    "    genes.loc[genes[\"strand\"] == \"+\", \"end\"] = genes.loc[\n",
    "        genes[\"strand\"] == \"+\", \"right\"\n",
    "    ].values\n",
    "\n",
    "    genes.loc[genes[\"strand\"] == \"-\", \"start\"] = genes.loc[\n",
    "        genes[\"strand\"] == \"-\", \"right\"\n",
    "    ].values\n",
    "\n",
    "    genes.loc[genes[\"strand\"] == \"-\", \"end\"] = genes.loc[\n",
    "        genes[\"strand\"] == \"-\", \"left\"\n",
    "    ].values\n",
    "    \n",
    "    genes['gene'] = genes.attributes.apply(extract_attribute,attr_name='gene')\n",
    "    genes['locus_tag'] = genes.attributes.apply(extract_attribute,attr_name='locus_tag')\n",
    "    genes['gene_or_locus'] = genes['gene'].fillna(genes['locus_tag'])\n",
    "    genes['product'] = genes.attributes.apply(extract_attribute,attr_name='product')\n",
    "    genes.loc[genes[\"type\"] == \"repeat_region\", \"gene\"] = \"REP\"\n",
    "    \n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene_or_locus</th>\n",
       "      <th>product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP024649.1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>517</td>\n",
       "      <td>1878</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxr...</td>\n",
       "      <td>517</td>\n",
       "      <td>1878</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>SaO11_00001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP024649.1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2158</td>\n",
       "      <td>3291</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-ATV02828.1;Parent=gene-SaO11_00002;Dbxr...</td>\n",
       "      <td>2158</td>\n",
       "      <td>3291</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>SaO11_00002</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>DNA polymerase III%2C beta chain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CP024649.1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3681</td>\n",
       "      <td>3917</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-ATV02829.1;Parent=gene-SaO11_00003;Dbxr...</td>\n",
       "      <td>3681</td>\n",
       "      <td>3917</td>\n",
       "      <td>None</td>\n",
       "      <td>SaO11_00003</td>\n",
       "      <td>SaO11_00003</td>\n",
       "      <td>RNA-binding protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CP024649.1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3968</td>\n",
       "      <td>5026</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-ATV02830.1;Parent=gene-SaO11_00004;Dbxr...</td>\n",
       "      <td>3968</td>\n",
       "      <td>5026</td>\n",
       "      <td>recF</td>\n",
       "      <td>SaO11_00004</td>\n",
       "      <td>recF</td>\n",
       "      <td>DNA replication and repair protein RecF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CP024649.1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5036</td>\n",
       "      <td>6970</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-ATV02831.1;Parent=gene-SaO11_00005;Dbxr...</td>\n",
       "      <td>5036</td>\n",
       "      <td>6970</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>SaO11_00005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>DNA gyrase subunit B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_id   source type  start   end score strand phase  \\\n",
       "2   CP024649.1  Genbank  CDS    517  1878     .      +     0   \n",
       "4   CP024649.1  Genbank  CDS   2158  3291     .      +     0   \n",
       "6   CP024649.1  Genbank  CDS   3681  3917     .      +     0   \n",
       "8   CP024649.1  Genbank  CDS   3968  5026     .      +     0   \n",
       "10  CP024649.1  Genbank  CDS   5036  6970     .      +     0   \n",
       "\n",
       "                                           attributes  left  right  gene  \\\n",
       "2   ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxr...   517   1878  dnaA   \n",
       "4   ID=cds-ATV02828.1;Parent=gene-SaO11_00002;Dbxr...  2158   3291  dnaN   \n",
       "6   ID=cds-ATV02829.1;Parent=gene-SaO11_00003;Dbxr...  3681   3917  None   \n",
       "8   ID=cds-ATV02830.1;Parent=gene-SaO11_00004;Dbxr...  3968   5026  recF   \n",
       "10  ID=cds-ATV02831.1;Parent=gene-SaO11_00005;Dbxr...  5036   6970  gyrB   \n",
       "\n",
       "      locus_tag gene_or_locus                                         product  \n",
       "2   SaO11_00001          dnaA  Chromosomal replication initiator protein DnaA  \n",
       "4   SaO11_00002          dnaN                DNA polymerase III%2C beta chain  \n",
       "6   SaO11_00003   SaO11_00003                             RNA-binding protein  \n",
       "8   SaO11_00004          recF         DNA replication and repair protein RecF  \n",
       "10  SaO11_00005          gyrB                            DNA gyrase subunit B  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_path = os.path.join(data_path, \"GCA_000189435.3_ASM18943v3_genomic.gff\")\n",
    "annotation=get_genome_annotations(gff_path)\n",
    "genes=get_genes_from_annotation(annotation)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "Y_RANGE = (-2, 2)\n",
    "def get_y_range() -> tuple:\n",
    "    \"\"\"Accessor that returns the Y range for the genome browser plot\n",
    "    \"\"\"\n",
    "    return Y_RANGE\n",
    "\n",
    "\n",
    "def get_all_glyphs(genes,bounds:tuple):\n",
    "    all_glyphs=get_gene_patches(genes, bounds[0], bounds[1])\n",
    "\n",
    "    ks=list(all_glyphs.keys())\n",
    "    ref_list_ix=ks.index('xs')\n",
    "    # Sort all the lists in the dictionary based on the values of the reference list\n",
    "    sorted_lists = sorted(zip(*[all_glyphs[k] for k in ks]), key= lambda x: x[ref_list_ix][0])\n",
    "\n",
    "    # Convert the sorted tuples back into separate lists\n",
    "    unzipped_lists = zip(*sorted_lists)\n",
    "\n",
    "    # Create a new dictionary with the same keys as the original dictionary, but with the sorted lists as values\n",
    "    all_glyphs = {k: list(t) for k, t in zip(ks, unzipped_lists)}\n",
    "    \n",
    "    return all_glyphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "def rect_patch(genes_region):\n",
    "    y_min, y_max = gene_y_range\n",
    "    xs = list(\n",
    "        zip(\n",
    "            genes_region.start.values,\n",
    "            genes_region.start.values,\n",
    "            genes_region.end.values,\n",
    "            genes_region.end.values,\n",
    "        )\n",
    "    )\n",
    "    xs = [np.array(x) for x in xs]\n",
    "    ys = [np.array([y_min, y_max, y_max, y_min]) for i in range(genes_region.shape[0])]\n",
    "    genes_mid = genes_region.left + (genes_region.right - genes_region.left) / 2\n",
    "    pos = list(genes_mid.values)\n",
    "    names = list(genes_region.gene.values)\n",
    "    product = list(genes_region[\"product\"].values)\n",
    "    color = [\"grey\"] * genes_region.shape[0]\n",
    "    return dict(\n",
    "        xs=xs,\n",
    "        ys=ys,\n",
    "        pos=pos,\n",
    "        names=[\"\"] * genes_region.shape[0],\n",
    "        gene=list(genes_region.gene.values),\n",
    "        locus_tag=list(genes_region.locus_tag.values),\n",
    "        hover_names=names,\n",
    "        product=product,\n",
    "        color=color,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "def arrow_patch(genes_region):\n",
    "    arr_plus = get_arrow_patch(genes_region[genes_region[\"strand\"] == \"+\"], \"+\")\n",
    "    arr_minus = get_arrow_patch(genes_region[genes_region[\"strand\"] == \"-\"], \"-\")\n",
    "    return dict([(k, arr_plus[k] + arr_minus[k]) for k in arr_plus.keys()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "gene_y_range = (-1.5, -1)\n",
    "\n",
    "def get_arrow_patch(genes_region, ori=\"+\"):\n",
    "    y_min, y_max = gene_y_range\n",
    "    y_min = y_min \n",
    "    if ori == \"+\":\n",
    "        xs = list(\n",
    "            zip(\n",
    "                genes_region.start.values,\n",
    "                genes_region.start.values,\n",
    "                np.maximum(genes_region.start.values, genes_region.end.values - 100),\n",
    "                genes_region.end.values,\n",
    "                np.maximum(genes_region.start.values, genes_region.end.values - 100),\n",
    "            )\n",
    "        )\n",
    "        color = [\"orange\"] * genes_region.shape[0]\n",
    "    elif ori == \"-\":\n",
    "        xs = list(\n",
    "            zip(\n",
    "                genes_region.start.values,\n",
    "                genes_region.start.values,\n",
    "                np.minimum(genes_region.start.values, genes_region.end.values + 100),\n",
    "                genes_region.end.values,\n",
    "                np.minimum(genes_region.start.values, genes_region.end.values + 100),\n",
    "            )\n",
    "        )\n",
    "        color = [\"purple\"] * genes_region.shape[0]\n",
    "\n",
    "    ys = [\n",
    "        np.array([y_min, y_max, y_max, (y_max + y_min) / 2, y_min])\n",
    "        for i in range(genes_region.shape[0])\n",
    "    ]\n",
    "    genes_mid = (genes_region.right + genes_region.left) / 2\n",
    "    pos = list(genes_mid.values)\n",
    "    return dict(\n",
    "        xs=xs,\n",
    "        ys=ys,\n",
    "        pos=pos,\n",
    "        names=list(genes_region.gene_or_locus.values),\n",
    "        gene=list(genes_region.gene.values),\n",
    "        locus_tag=list(genes_region.locus_tag.values),\n",
    "        hover_names=list(genes_region.gene_or_locus.values),\n",
    "        product=list(genes_region[\"product\"].values),\n",
    "        color=color,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "def get_gene_patches(genes, left, right):\n",
    "    genes_region = genes[\n",
    "        (genes[\"right\"] > left)\n",
    "        & (genes[\"left\"] < right)\n",
    "        & (genes[\"type\"] != \"repeat_region\")\n",
    "    ]\n",
    "    arr = arrow_patch(genes_region)\n",
    "    # repeat_region\n",
    "    rep_region = genes[\n",
    "        (genes[\"right\"] > left)\n",
    "        & (genes[\"left\"] < right)\n",
    "        & (genes[\"type\"] == \"repeat_region\")\n",
    "    ]\n",
    "    rect = rect_patch(rep_region)\n",
    "\n",
    "    # concatenate patches\n",
    "    res = dict([(k, arr[k] + rect[k]) for k in arr.keys()])\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': [(517, 517, 1778, 1878, 1778),\n",
       "  (2158, 2158, 3191, 3291, 3191),\n",
       "  (3681, 3681, 3817, 3917, 3817),\n",
       "  (3968, 3968, 4926, 5026, 4926)],\n",
       " 'ys': [array([-1.5 , -1.  , -1.  , -1.25, -1.5 ]),\n",
       "  array([-1.5 , -1.  , -1.  , -1.25, -1.5 ]),\n",
       "  array([-1.5 , -1.  , -1.  , -1.25, -1.5 ]),\n",
       "  array([-1.5 , -1.  , -1.  , -1.25, -1.5 ])],\n",
       " 'pos': [1197.5, 2724.5, 3799.0, 4497.0],\n",
       " 'names': ['dnaA', 'dnaN', 'SaO11_00003', 'recF'],\n",
       " 'gene': ['dnaA', 'dnaN', None, 'recF'],\n",
       " 'locus_tag': ['SaO11_00001', 'SaO11_00002', 'SaO11_00003', 'SaO11_00004'],\n",
       " 'hover_names': ['dnaA', 'dnaN', 'SaO11_00003', 'recF'],\n",
       " 'product': ['Chromosomal replication initiator protein DnaA',\n",
       "  'DNA polymerase III%2C beta chain',\n",
       "  'RNA-binding protein',\n",
       "  'DNA replication and repair protein RecF'],\n",
       " 'color': ['orange', 'orange', 'orange', 'orange']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "get_gene_patches(genes,0,5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
