{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# utils\n",
    "\n",
    "> This contains useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import io\n",
    "\n",
    "from bokeh.plotting import figure\n",
    "from bokeh.models.tools import BoxZoomTool\n",
    "from bokeh.models import HoverTool, NumeralTickFormatter, LabelSet\n",
    "from bokeh.models.glyphs import Patches\n",
    "from bokeh.models import (\n",
    "    CustomJS,\n",
    "    Range1d,\n",
    "    ColumnDataSource,\n",
    ")\n",
    "\n",
    "from collections import defaultdict\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def parse_gff(gff_path)->pd.DataFrame:\n",
    "    cwd = os.getcwd()\n",
    "    with open(gff_path,\"r\") as gff_file:\n",
    "        # Create an in-memory file buffer using the io.StringIO class\n",
    "        file_buffer = io.StringIO()\n",
    "        for line in gff_file:\n",
    "            if line[0]==\"#\":\n",
    "                continue\n",
    "            else:\n",
    "                # Write each line to the file buffer\n",
    "                file_buffer.write(line)\n",
    "                \n",
    "        # Reset the file pointer to the beginning of the file buffer\n",
    "        file_buffer.seek(0)       \n",
    "\n",
    "        df=pd.read_csv(file_buffer,sep=\"\\t\",header=None)\n",
    "        df.columns=[\"seq_id\", \"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#| export\n",
    "from genomenotebook.data import get_example_data_dir\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...  \n",
       "1  ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...  \n",
       "2  ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...  \n",
       "3  ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...  \n",
       "4  ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "df=parse_gff(gff_path)\n",
    "assert type(df)==pd.DataFrame\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def set_positions(annotation: pd.DataFrame, # an annotation DataFrame extracted from a gff file\n",
    "                            ):\n",
    "    \"\"\"Sets left and right as the position of the feature on the sequence, left is always lower than right.\n",
    "    start and end represent the begining and end of the feature where start can be greater than end depending on the feature strand.\n",
    "    \"\"\"\n",
    "    annotation=annotation.copy()\n",
    "    annotation.loc[:, \"left\"] = annotation[[\"start\"]].values\n",
    "    annotation.loc[:, \"right\"] = annotation[[\"end\"]].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"+\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"left\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"right\"].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"-\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"right\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"left\"].values\n",
    "    \n",
    "    annotation[\"middle\"] = (annotation.right + annotation.left) / 2\n",
    "    \n",
    "    return annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>2320826.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  left    right     middle  \n",
       "0  ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...     1  4641652  2320826.5  \n",
       "1  ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...   190      255      222.5  \n",
       "2  ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...   190      255      222.5  \n",
       "3  ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...   337     2799     1568.0  \n",
       "4  ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...   337     2799     1568.0  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set_positions(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_genome_annotations(gff_path: str, seq_id: str=None):\n",
    "    \"\"\"Parses a gff file and extracts the features that belong to seq_id. Raises a warning if resulting DataFrame is empty.\"\"\"\n",
    "    annotation = parse_gff(gff_path)\n",
    "    if seq_id:\n",
    "        annotation = annotation.loc[(annotation.seq_id == seq_id)]\n",
    "        if len(annotation)==0:\n",
    "            warnings.warn(\"The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\")\n",
    "    \n",
    "    annotation = set_positions(annotation)\n",
    "    return annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>2320826.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  left    right     middle  \n",
       "0  ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...     1  4641652  2320826.5  \n",
       "1  ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...   190      255      222.5  \n",
       "2  ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...   190      255      222.5  \n",
       "3  ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...   337     2799     1568.0  \n",
       "4  ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...   337     2799     1568.0  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "annotation=get_genome_annotations(gff_path, \"U00096.3\")\n",
    "annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4291/114227395.py:8: UserWarning: The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\n",
      "  warnings.warn(\"The annotation DataFrame is empty. Check that the fasta and gff files have the same sequence id\")\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "#Testing mistake in seq_id\n",
    "annotation=get_genome_annotations(gff_path, \"U00097.3\") #mistake in seq_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_attribute(input_str:str, #attribute string to parse\n",
    "                      attr_name:str, #name of the attribute to extract\n",
    "                     ) -> str:\n",
    "    \"\"\"Extracts the attribute called attr_name from the GFF attributes string\"\"\"\n",
    "    \n",
    "    pattern = f\"[{attr_name[0].lower()}{attr_name[0].upper()}]{attr_name[1:]}=(?P<{attr_name}>[^;]+)\"\n",
    "    match = re.search(pattern, input_str)\n",
    "    if match:\n",
    "        return match.groupdict()[attr_name]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dnaA'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_str = 'ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxref=NCBI_GP:ATV02827.1;Name=ATV02827.1;gbkey=CDS;gene=dnaA;locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "extract_attribute(input_str,\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2        AAC73112.1\n",
       "4        AAC73113.1\n",
       "6        AAC73114.1\n",
       "8        AAC73115.1\n",
       "10       AAC73116.1\n",
       "            ...    \n",
       "10112    AAC77352.1\n",
       "10114    AAC77353.1\n",
       "10116    AAC77354.1\n",
       "10118    AAC77355.1\n",
       "10120    AAC77356.1\n",
       "Name: attributes, Length: 5256, dtype: object"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation=get_genome_annotations(gff_path, \"U00096.3\")\n",
    "genes = annotation[annotation.type.isin([\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"])    ]\n",
    "genes.attributes.apply(extract_attribute,attr_name='protein_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_all_attributes(input_str:str)->dict:\n",
    "    \"\"\"Extracts all attributes from the GFF attributes column\"\"\"\n",
    "    \n",
    "    pattern = \"(?P<key>\\w+[-\\w]*)=(?P<value>[^;]+)\"\n",
    "    match = re.findall(pattern, input_str)\n",
    "    d=defaultdict()\n",
    "    d.update(match)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(None,\n",
       "            {'ID': 'cds-ATV02827.1',\n",
       "             'Parent': 'gene-SaO11_00001',\n",
       "             'Dbxref': 'NCBI_GP:ATV02827.1',\n",
       "             'Name': 'ATV02827.1',\n",
       "             'gbkey': 'CDS',\n",
       "             'gene': 'dnaA',\n",
       "             'locus_tag': 'SaO11_00001',\n",
       "             'product': 'Chromosomal replication initiator protein DnaA',\n",
       "             'protein_id': 'ATV02827.1',\n",
       "             'transl_table': '11'})"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_all_attributes(input_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_feature_types(gff_path):\n",
    "    features=parse_gff(gff_path)\n",
    "    all_keys=set(features.type.values)\n",
    "    return all_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CDS',\n",
       " 'exon',\n",
       " 'gene',\n",
       " 'mobile_genetic_element',\n",
       " 'ncRNA',\n",
       " 'origin_of_replication',\n",
       " 'pseudogene',\n",
       " 'rRNA',\n",
       " 'recombination_feature',\n",
       " 'region',\n",
       " 'repeat_region',\n",
       " 'sequence_feature',\n",
       " 'tRNA'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "available_feature_types(gff_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_attributes(gff_path):\n",
    "    features=parse_gff(gff_path)\n",
    "    attr_dicts=features.attributes.apply(extract_all_attributes)\n",
    "    all_keys=list(set().union(*[d.keys() for d in attr_dicts]))\n",
    "    return all_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['transl_except', 'mobile_element_type', 'gene_synonym', 'mol_type', 'product', 'Note', 'gene_biotype', 'Is_circular', 'part', 'gbkey', 'orig_transcript_id', 'Parent', 'gene', 'rpt_type', 'pseudo', 'substrain', 'recombination_class', 'protein_id', 'transl_table', 'orig_protein_id', 'strain', 'locus_tag', 'ID', 'Dbxref', 'genome', 'Name', 'exception']\n"
     ]
    }
   ],
   "source": [
    "print(available_attributes(gff_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def attributes_to_columns(features: pd.DataFrame):\n",
    "    attr_dicts=features.attributes.apply(extract_all_attributes)\n",
    "    all_keys=list(set().union(*[d.keys() for d in attr_dicts]))\n",
    "    \n",
    "    attr_dict=dict([(k,[d.get(k,None) for d in attr_dicts]) for k in all_keys])\n",
    "    features=features.copy()\n",
    "    for k,v in attr_dict.items():\n",
    "        features[k]=v\n",
    "        \n",
    "    return features\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>...</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>genome</th>\n",
       "      <th>Is_circular</th>\n",
       "      <th>strain</th>\n",
       "      <th>Name</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>true</td>\n",
       "      <td>K-12</td>\n",
       "      <td>ANONYMOUS</td>\n",
       "      <td>Src</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>U00096.3:1..4641652</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>thrL</td>\n",
       "      <td>Gene</td>\n",
       "      <td>b0001</td>\n",
       "      <td>None</td>\n",
       "      <td>gene-b0001</td>\n",
       "      <td>thrL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>AAC73112.1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC73112.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>b0001</td>\n",
       "      <td>11</td>\n",
       "      <td>cds-AAC73112.1</td>\n",
       "      <td>thrL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...</td>\n",
       "      <td>337</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>b0002</td>\n",
       "      <td>None</td>\n",
       "      <td>gene-b0002</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "      <td>337</td>\n",
       "      <td>...</td>\n",
       "      <td>AAC73113.1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC73113.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>b0002</td>\n",
       "      <td>11</td>\n",
       "      <td>cds-AAC73113.1</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  left  ...  protein_id  \\\n",
       "0  ID=U00096.3:1..4641652;Dbxref=taxon:511145;Is_...     1  ...        None   \n",
       "1  ID=gene-b0001;Dbxref=ASAP:ABE-0000006,ECOCYC:E...   190  ...        None   \n",
       "2  ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...   190  ...  AAC73112.1   \n",
       "3  ID=gene-b0002;Dbxref=ASAP:ABE-0000008,ECOCYC:E...   337  ...        None   \n",
       "4  ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...   337  ...  AAC73113.1   \n",
       "\n",
       "       genome Is_circular strain        Name gbkey locus_tag transl_table  \\\n",
       "0  chromosome        true   K-12   ANONYMOUS   Src      None         None   \n",
       "1        None        None   None        thrL  Gene     b0001         None   \n",
       "2        None        None   None  AAC73112.1   CDS     b0001           11   \n",
       "3        None        None   None        thrA  Gene     b0002         None   \n",
       "4        None        None   None  AAC73113.1   CDS     b0002           11   \n",
       "\n",
       "                    ID  gene  \n",
       "0  U00096.3:1..4641652  None  \n",
       "1           gene-b0001  thrL  \n",
       "2       cds-AAC73112.1  thrL  \n",
       "3           gene-b0002  thrA  \n",
       "4       cds-AAC73113.1  thrA  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attributes_to_columns(annotation.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "input_str = 'locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "assert extract_attribute(input_str,\"gene\") == None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "attributes=[\"gene\", \"locus_tag\", \"product\"]\n",
    "def get_features_from_annotation(annotation: pd.DataFrame, #annotation DataFrame extracted from GFF file\n",
    "                                 feature_types: list = [\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"], # list of feature types to extract\n",
    "                                 attributes: list = attributes # list of attributes to extract\n",
    "                                ):\n",
    "    \"\"\"Filters annotations to keep only features of type `feature_type`.\n",
    "       Extracts each attribute from `attributes` as a column.\n",
    "    \"\"\"\n",
    "    features = annotation[\n",
    "        annotation.type.isin(feature_types)\n",
    "    ].copy()\n",
    " \n",
    "    for attr in attributes:\n",
    "        features[attr] = features.attributes.apply(extract_attribute,attr_name=attr)\n",
    "        \n",
    "    features.loc[features.type == \"repeat_region\", \"gene\"] = \"REP\"\n",
    "    features['gene_or_locus'] = features['gene'].fillna(features[attributes[0]])\n",
    "    \n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>gene_or_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "      <td>thrL</td>\n",
       "      <td>b0001</td>\n",
       "      <td>thr operon leader peptide</td>\n",
       "      <td>thrL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "      <td>thrA</td>\n",
       "      <td>b0002</td>\n",
       "      <td>fused aspartate kinase/homoserine dehydrogenase 1</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73114.1;Parent=gene-b0003;Dbxref=Uni...</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>3267.0</td>\n",
       "      <td>thrB</td>\n",
       "      <td>b0003</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>thrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3734</td>\n",
       "      <td>5020</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73115.1;Parent=gene-b0004;Dbxref=Uni...</td>\n",
       "      <td>3734</td>\n",
       "      <td>5020</td>\n",
       "      <td>4377.0</td>\n",
       "      <td>thrC</td>\n",
       "      <td>b0004</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>thrC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-AAC73116.1;Parent=gene-b0005;Dbxref=Uni...</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>5382.0</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>b0005</td>\n",
       "      <td>DUF2502 domain-containing protein YaaX</td>\n",
       "      <td>yaaX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_id   source type  start   end score strand phase  \\\n",
       "2   U00096.3  Genbank  CDS    190   255     .      +     0   \n",
       "4   U00096.3  Genbank  CDS    337  2799     .      +     0   \n",
       "6   U00096.3  Genbank  CDS   2801  3733     .      +     0   \n",
       "8   U00096.3  Genbank  CDS   3734  5020     .      +     0   \n",
       "10  U00096.3  Genbank  CDS   5234  5530     .      +     0   \n",
       "\n",
       "                                           attributes  left  right  middle  \\\n",
       "2   ID=cds-AAC73112.1;Parent=gene-b0001;Dbxref=Uni...   190    255   222.5   \n",
       "4   ID=cds-AAC73113.1;Parent=gene-b0002;Dbxref=Uni...   337   2799  1568.0   \n",
       "6   ID=cds-AAC73114.1;Parent=gene-b0003;Dbxref=Uni...  2801   3733  3267.0   \n",
       "8   ID=cds-AAC73115.1;Parent=gene-b0004;Dbxref=Uni...  3734   5020  4377.0   \n",
       "10  ID=cds-AAC73116.1;Parent=gene-b0005;Dbxref=Uni...  5234   5530  5382.0   \n",
       "\n",
       "    gene locus_tag                                            product  \\\n",
       "2   thrL     b0001                          thr operon leader peptide   \n",
       "4   thrA     b0002  fused aspartate kinase/homoserine dehydrogenase 1   \n",
       "6   thrB     b0003                                  homoserine kinase   \n",
       "8   thrC     b0004                                 threonine synthase   \n",
       "10  yaaX     b0005             DUF2502 domain-containing protein YaaX   \n",
       "\n",
       "   gene_or_locus  \n",
       "2           thrL  \n",
       "4           thrA  \n",
       "6           thrB  \n",
       "8           thrC  \n",
       "10          yaaX  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features=get_features_from_annotation(annotation)\n",
    "features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "gene_y_range = (-1.5, -1)\n",
    "\n",
    "def arrow_coordinates(feature):\n",
    "    feature_size = feature.right - feature.left\n",
    "    \n",
    "    if feature.strand==\"+\":\n",
    "        arrow_base = feature.end - np.minimum(feature_size, 100)\n",
    "    else:\n",
    "        arrow_base = feature.end + np.minimum(feature_size, 100)\n",
    "    \n",
    "    xs=(feature.start,\n",
    "        feature.start,\n",
    "        arrow_base,\n",
    "        feature.end,\n",
    "        arrow_base\n",
    "       )\n",
    "    \n",
    "    y_min, y_max = gene_y_range\n",
    "    ys = (y_min, y_max, y_max, (y_max + y_min) / 2, y_min)\n",
    "    return xs, ys\n",
    "\n",
    "def box_coordinates(feature):\n",
    "    xs=(feature.left, feature.left,\n",
    "        feature.right, feature.right)\n",
    "    y_min, y_max = gene_y_range\n",
    "    ys = (y_min, y_max, y_max, y_min)\n",
    "    return xs, ys\n",
    "\n",
    "default_glyphs=defaultdict(lambda: (\"arrow\",(\"purple\",\"orange\"))) #the default glyph will be the same as for CDS etc.\n",
    "default_glyphs.update(dict([(f,(\"arrow\",(\"purple\",\"orange\"))) for f in [\"CDS\", \"ncRNA\", \"rRNA\", \"tRNA\"]]))\n",
    "default_glyphs['repeat_region']=(\"box\",(\"grey\",))\n",
    "\n",
    "def get_patch_coordinates(feature, # row of a pandas DataFrame extracted from a GFF file\n",
    "                          patch_dict: dict = default_glyphs # a dictionnary containing as key a feature type and as value a patch definition.\n",
    "                         ):\n",
    "    \"\"\"\n",
    "    patchs are defined with a patch a tuple: (patch_type, (patch_color_plus, patch_color_minus)). \n",
    "    Different colors can be specified depending on the strand.\"\"\"\n",
    "    \n",
    "    patch_type, patch_colors = patch_dict[feature.type]\n",
    "    if len(patch_colors)>1:\n",
    "        color_dic={\"+\":patch_colors[0],\n",
    "                   \"-\":patch_colors[1]}\n",
    "    else:\n",
    "        color_dic=defaultdict(lambda: patch_colors[0])\n",
    "        \n",
    "    if patch_type==\"arrow\":\n",
    "        return arrow_coordinates(feature), color_dic[feature.strand]\n",
    "    elif patch_type==\"box\":\n",
    "        return box_coordinates(feature), color_dic[feature.strand]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((((190, 190, 190, 255, 190), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((337, 337, 2699, 2799, 2699), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((2801, 2801, 3633, 3733, 3633), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((3734, 3734, 4920, 5020, 4920), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((5234, 5234, 5430, 5530, 5430), (-1.5, -1, -1, -1.25, -1.5))),\n",
       " ('purple', 'purple', 'purple', 'purple', 'purple'))"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates, colors = zip(*features.apply(get_patch_coordinates,axis=1))\n",
    "coordinates[:5], colors[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_feature_patches(features: pd.DataFrame, #DataFrame of the features \n",
    "                        left: int, #left limit\n",
    "                        right: int, #right limit\n",
    "                        patch_dict: dict = default_glyphs, #glyphs to use for each feature type\n",
    "                        attributes: list = attributes, #list of attributes to display when hovering\n",
    "                        name: str = attributes[0] #attribute to be displayed as the feature name\n",
    "                       ):\n",
    "    features=features.loc[(features[\"right\"] > left) & (features[\"left\"] < right)]\n",
    "    \n",
    "    if len(features)>0:\n",
    "        coordinates, colors = zip(*features.apply(get_patch_coordinates,patch_dict=patch_dict,axis=1))\n",
    "        xs, ys = zip(*coordinates)\n",
    "    else:\n",
    "        colors = []\n",
    "        xs, ys = [], []\n",
    "    names = list(features[name].fillna(\"\").values) #list(features.gene.fillna(features[\"locus_tag\"]).values)\n",
    "    out=dict(xs=xs,\n",
    "             ys=ys,\n",
    "             color=colors,\n",
    "             pos=features.middle.values,\n",
    "             names=names,\n",
    "             hover_names=names,\n",
    "            )\n",
    "\n",
    "    out.update(features[attributes].to_dict(orient='list'))\n",
    "    return out\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(<function __main__.<lambda>()>,\n",
       "            {'CDS': ('arrow', ('purple', 'orange')),\n",
       "             'ncRNA': ('arrow', ('purple', 'orange')),\n",
       "             'rRNA': ('arrow', ('purple', 'orange')),\n",
       "             'tRNA': ('arrow', ('purple', 'orange')),\n",
       "             'repeat_region': ('box', ('grey',))})"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "default_glyphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': ((10494, 10494, 10028, 9928, 10028),\n",
       "  (11356, 11356, 10743, 10643, 10743),\n",
       "  (10830, 10830, 11215, 11315, 11215),\n",
       "  (11786, 11786, 11482, 11382, 11482),\n",
       "  (12163, 12163, 13979, 14079, 13979),\n",
       "  (14168, 14168, 15198, 15298, 15198)),\n",
       " 'ys': ((-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5)),\n",
       " 'color': ('orange', 'orange', 'purple', 'orange', 'purple', 'purple'),\n",
       " 'pos': array([10211. , 10999.5, 11072.5, 11584. , 13121. , 14733. ]),\n",
       " 'names': ['satP', 'yaaW', 'mbiA', 'yaaI', 'dnaK', 'dnaJ'],\n",
       " 'hover_names': ['satP', 'yaaW', 'mbiA', 'yaaI', 'dnaK', 'dnaJ'],\n",
       " 'gene': ['satP', 'yaaW', 'mbiA', 'yaaI', 'dnaK', 'dnaJ'],\n",
       " 'locus_tag': ['b0010', 'b0011', 'b0012', 'b0013', 'b0014', 'b0015'],\n",
       " 'product': ['acetate/succinate:H(+) symporter',\n",
       "  'putative enzyme-specific chaperone YaaW',\n",
       "  'uncharacterized protein MbiA',\n",
       "  'DUF2541 domain-containing protein YaaI',\n",
       "  'chaperone protein DnaK',\n",
       "  'chaperone protein DnaJ']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_patches(features,10000,15000, patch_dict=default_glyphs, name=\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sort_list_dict(d: dict[list], #a dictionnary for which all values are of type list\n",
    "                   ref_list=\"xs\", #key of the list to use as a reference for sorting\n",
    "                   func=lambda x: x, #A custom function can be supplied to customize the sort order. Default is the identity function.\n",
    "                  ):\n",
    "    ks=list(d.keys())\n",
    "    ref_list_ix=ks.index(ref_list)\n",
    "    # Sort all the lists in the dictionary based on the values of the reference list\n",
    "    sorted_lists = sorted(zip(*[d[k] for k in ks]), key= lambda x: func(x[ref_list_ix]))\n",
    "\n",
    "    # Convert the sorted tuples back into separate lists\n",
    "    unzipped_lists = zip(*sorted_lists)\n",
    "\n",
    "    # Create a new dictionary with the same keys as the original dictionary, but with the sorted lists as values\n",
    "    d = {k: list(t) for k, t in zip(ks, unzipped_lists)}\n",
    "    return d\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': [[3, 1], [2, 5]], 'a': [3, 1], 'b': ['d', 'c']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d={'xs':[[2,5],[3,1]],'a':[1,3],'b':[\"c\",\"d\"]}\n",
    "sort_list_dict(d, ref_list='xs', func= lambda x: x[1]) #sort according to the second element of each element of list 'xs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "Y_RANGE = (-2, 2)\n",
    "def get_y_range() -> tuple:\n",
    "    \"\"\"Accessor that returns the Y range for the genome browser plot\n",
    "    \"\"\"\n",
    "    return Y_RANGE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "def get_all_glyphs(genes,bounds:tuple) -> dict:\n",
    "    all_glyphs=get_gene_patches(genes, bounds[0], bounds[1])\n",
    "\n",
    "    ks=list(all_glyphs.keys())\n",
    "    ref_list_ix=ks.index('xs')\n",
    "    # Sort all the lists in the dictionary based on the values of the reference list\n",
    "    sorted_lists = sorted(zip(*[all_glyphs[k] for k in ks]), key= lambda x: x[ref_list_ix][0])\n",
    "\n",
    "    # Convert the sorted tuples back into separate lists\n",
    "    unzipped_lists = zip(*sorted_lists)\n",
    "\n",
    "    # Create a new dictionary with the same keys as the original dictionary, but with the sorted lists as values\n",
    "    all_glyphs = {k: list(t) for k, t in zip(ks, unzipped_lists)}\n",
    "    \n",
    "    return all_glyphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def create_genome_browser_plot(glyphSource, x_range, attributes=attributes, **kwargs):\n",
    "    plot_height = kwargs.get(\"plot_height\", 150)\n",
    "    label_angle = kwargs.get(\"label_angle\", 45)\n",
    "    text_font_size = kwargs.get(\"text_font_size\", \"10pt\")\n",
    "    output_backend = kwargs.get(\"output_backend\", \"webgl\")\n",
    "    \n",
    "    y_min, y_max = get_y_range()\n",
    "    p_annot = figure(\n",
    "        tools = \"xwheel_zoom,xpan,save\",\n",
    "        active_scroll = \"xwheel_zoom\",\n",
    "        height = plot_height,\n",
    "        x_range = x_range,\n",
    "        y_range = Range1d(y_min, y_max),\n",
    "        output_backend=output_backend,\n",
    "    )\n",
    "    # Add tool\n",
    "    p_annot.add_tools(BoxZoomTool(dimensions=\"width\"))\n",
    "\n",
    "    # Format x axis values\n",
    "    p_annot.xaxis[0].formatter = NumeralTickFormatter(format=\"0,0\")\n",
    "    # Hide grid\n",
    "    p_annot.xgrid.visible = False\n",
    "    p_annot.ygrid.visible = False\n",
    "    # Hide axis\n",
    "    p_annot.yaxis.visible = False\n",
    "    glyph = p_annot.add_glyph(\n",
    "        glyphSource, Patches(xs=\"xs\", ys=\"ys\", fill_color=\"color\")\n",
    "    )\n",
    "    # gene labels in the annotation track\n",
    "    # This seems to be necessary to show the labels\n",
    "    p_annot.scatter(x=\"pos\", y=0, size=0, source=glyphSource)\n",
    "    labels = LabelSet(\n",
    "        x=\"pos\",\n",
    "        y=-0.9,\n",
    "        text=\"names\",\n",
    "        level=\"glyph\",\n",
    "        angle=label_angle,\n",
    "        text_font_size=text_font_size,\n",
    "        x_offset=-5,\n",
    "        y_offset=0,\n",
    "        source=glyphSource,\n",
    "        text_align='left',\n",
    "    )\n",
    "\n",
    "    p_annot.add_layout(labels)\n",
    "    tooltips=[(attr,f\"@{attr}\") for attr in attributes]\n",
    "    p_annot.add_tools(\n",
    "        HoverTool(\n",
    "            renderers=[glyph],\n",
    "            tooltips=tooltips,\n",
    "        )\n",
    "    )\n",
    "    return p_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.models import (\n",
    "    Range1d,\n",
    "    ColumnDataSource,\n",
    ")\n",
    "from bokeh.plotting import show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"b7b4a7b6-4202-4c70-8a0f-4b973f0f9bf4\" data-root-id=\"p1006\" style=\"display: contents;\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "  const docs_json = {\"04375ceb-eb8e-4d0f-8f05-e30c378c0bd9\":{\"version\":\"3.1.0\",\"title\":\"Bokeh Application\",\"defs\":[],\"roots\":[{\"type\":\"object\",\"name\":\"Figure\",\"id\":\"p1006\",\"attributes\":{\"height\":150,\"x_range\":{\"type\":\"object\",\"name\":\"Range1d\",\"id\":\"p1001\",\"attributes\":{\"start\":2000,\"end\":4000,\"bounds\":[0,5000],\"min_interval\":40,\"max_interval\":100000}},\"y_range\":{\"type\":\"object\",\"name\":\"Range1d\",\"id\":\"p1005\",\"attributes\":{\"start\":-2,\"end\":2}},\"x_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1017\"},\"y_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1019\"},\"title\":{\"type\":\"object\",\"name\":\"Title\",\"id\":\"p1013\"},\"renderers\":[{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1048\",\"attributes\":{\"data_source\":{\"type\":\"object\",\"name\":\"ColumnDataSource\",\"id\":\"p1002\",\"attributes\":{\"selected\":{\"type\":\"object\",\"name\":\"Selection\",\"id\":\"p1004\",\"attributes\":{\"indices\":[],\"line_indices\":[]}},\"selection_policy\":{\"type\":\"object\",\"name\":\"UnionRenderers\",\"id\":\"p1003\"},\"data\":{\"type\":\"map\",\"entries\":[[\"xs\",[[337,337,2699,2799,2699],[2801,2801,3633,3733,3633],[3734,3734,4920,5020,4920]]],[\"ys\",[[-1.5,-1,-1,-1.25,-1.5],[-1.5,-1,-1,-1.25,-1.5],[-1.5,-1,-1,-1.25,-1.5]]],[\"color\",[\"purple\",\"purple\",\"purple\"]],[\"pos\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"AAAAAACAmEAAAAAAAIapQAAAAAAAGbFA\"},\"shape\":[3],\"dtype\":\"float64\",\"order\":\"little\"}],[\"names\",[\"thrA\",\"thrB\",\"thrC\"]],[\"hover_names\",[\"thrA\",\"thrB\",\"thrC\"]],[\"gene\",[\"thrA\",\"thrB\",\"thrC\"]],[\"locus_tag\",[\"b0002\",\"b0003\",\"b0004\"]],[\"product\",[\"fused aspartate kinase/homoserine dehydrogenase 1\",\"homoserine kinase\",\"threonine synthase\"]]]}}},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1049\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1050\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Patches\",\"id\":\"p1047\",\"attributes\":{\"xs\":{\"type\":\"field\",\"field\":\"xs\"},\"ys\":{\"type\":\"field\",\"field\":\"ys\"},\"fill_color\":{\"type\":\"field\",\"field\":\"color\"}}}}},{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1057\",\"attributes\":{\"data_source\":{\"id\":\"p1002\"},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1058\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1059\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1054\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"}}},\"nonselection_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1055\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.1},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.1},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.1}}},\"muted_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1056\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.2},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.2},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.2}}}}}],\"toolbar\":{\"type\":\"object\",\"name\":\"Toolbar\",\"id\":\"p1014\",\"attributes\":{\"tools\":[{\"type\":\"object\",\"name\":\"WheelZoomTool\",\"id\":\"p1035\",\"attributes\":{\"dimensions\":\"width\"}},{\"type\":\"object\",\"name\":\"PanTool\",\"id\":\"p1036\",\"attributes\":{\"dimensions\":\"width\"}},{\"type\":\"object\",\"name\":\"SaveTool\",\"id\":\"p1037\"},{\"type\":\"object\",\"name\":\"BoxZoomTool\",\"id\":\"p1042\",\"attributes\":{\"dimensions\":\"width\",\"overlay\":{\"type\":\"object\",\"name\":\"BoxAnnotation\",\"id\":\"p1043\",\"attributes\":{\"syncable\":false,\"level\":\"overlay\",\"visible\":false,\"left_units\":\"canvas\",\"right_units\":\"canvas\",\"bottom_units\":\"canvas\",\"top_units\":\"canvas\",\"line_color\":\"black\",\"line_alpha\":1.0,\"line_width\":2,\"line_dash\":[4,4],\"fill_color\":\"lightgrey\",\"fill_alpha\":0.5}}}},{\"type\":\"object\",\"name\":\"HoverTool\",\"id\":\"p1064\",\"attributes\":{\"renderers\":[{\"id\":\"p1048\"}],\"tooltips\":[[\"gene\",\"@gene\"],[\"locus_tag\",\"@locus_tag\"],[\"product\",\"@product\"]]}}],\"active_scroll\":{\"id\":\"p1035\"}}},\"left\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1028\",\"attributes\":{\"visible\":false,\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1029\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1031\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1030\"}}}],\"below\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1021\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1022\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"NumeralTickFormatter\",\"id\":\"p1045\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1023\"}}}],\"center\":[{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1027\",\"attributes\":{\"visible\":false,\"axis\":{\"id\":\"p1021\"}}},{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1034\",\"attributes\":{\"visible\":false,\"dimension\":1,\"axis\":{\"id\":\"p1028\"}}},{\"type\":\"object\",\"name\":\"LabelSet\",\"id\":\"p1060\",\"attributes\":{\"level\":\"glyph\",\"source\":{\"id\":\"p1002\"},\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":-0.9},\"text\":{\"type\":\"field\",\"field\":\"names\"},\"angle\":{\"type\":\"value\",\"value\":45},\"x_offset\":{\"type\":\"value\",\"value\":-5},\"text_font_size\":{\"type\":\"value\",\"value\":\"10pt\"}}}],\"output_backend\":\"webgl\"}}],\"callbacks\":{\"type\":\"map\"}}};\n",
       "  const render_items = [{\"docid\":\"04375ceb-eb8e-4d0f-8f05-e30c378c0bd9\",\"roots\":{\"p1006\":\"b7b4a7b6-4202-4c70-8a0f-4b973f0f9bf4\"},\"root_ids\":[\"p1006\"]}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    let attempts = 0;\n",
       "    const timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "p1006"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "features = get_features_from_annotation(annotation)\n",
    "x_range = Range1d(\n",
    "            2000, 4000, \n",
    "            bounds=(0,5000), \n",
    "            max_interval=100000,\n",
    "            min_interval=40\n",
    "        )\n",
    "\n",
    "feature_patches = get_feature_patches(features, x_range.start, x_range.end)\n",
    "glyph_source = ColumnDataSource(feature_patches)\n",
    "p = create_genome_browser_plot(glyph_source, x_range)\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def split_string(string, max_length=10):\n",
    "    if len(string) <= max_length:\n",
    "        return string\n",
    "    else:\n",
    "        split_index = max_length\n",
    "        while split_index > 0 and string[split_index] != ' ':\n",
    "            split_index -= 1\n",
    "        if split_index == 0:\n",
    "            split_index = max_length  # If no suitable breaking point found, split at max_length\n",
    "        return string[:split_index] + '\\n' + split_string(string[split_index:].lstrip(), max_length)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a very long string that needs to be split\n",
      "into multiple lines because it exceeds 50\n",
      "characters.\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "long_string = \"This is a very long string that needs to be split into multiple lines because it exceeds 50 characters.\"\n",
    "\n",
    "split_result = split_string(long_string, max_length=50)\n",
    "print(split_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
