{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# utils\n",
    "\n",
    "> This contains useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import io\n",
    "\n",
    "from bokeh.plotting import figure\n",
    "from bokeh.models.tools import BoxZoomTool\n",
    "from bokeh.models import HoverTool, NumeralTickFormatter, LabelSet\n",
    "from bokeh.models.glyphs import Patches\n",
    "from bokeh.models import (\n",
    "    CustomJS,\n",
    "    Range1d,\n",
    "    ColumnDataSource,\n",
    ")\n",
    "\n",
    "from collections import defaultdict\n",
    "import warnings\n",
    "import gzip\n",
    "import urllib.request\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def is_gzipped_file(file_path):\n",
    "    try:\n",
    "        with gzip.open(file_path, 'rb') as f:\n",
    "            # Attempt to read a small chunk from the file\n",
    "            f.read(1)\n",
    "        return True\n",
    "    except IOError:\n",
    "        return False\n",
    "    \n",
    "def download_file(url, save_path):\n",
    "    if os.path.exists(save_path):\n",
    "        print(f\"File already exists: {save_path}\")\n",
    "    else:\n",
    "        urllib.request.urlretrieve(url, save_path)\n",
    "        print(f\"File downloaded and saved: {save_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists: GRCh38_latest_genomic.gff.gz\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example usage\n",
    "file_url = 'https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gff.gz'\n",
    "human_genome_gff = 'GRCh38_latest_genomic.gff.gz'\n",
    "\n",
    "download_file(file_url, human_genome_gff)\n",
    "is_gzipped_file(human_genome_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_attribute(input_str:str, #attribute string to parse\n",
    "                      attr_name:str, #name of the attribute to extract\n",
    "                     ) -> str:\n",
    "    \"\"\"Extracts the attribute called attr_name from the GFF attributes string\"\"\"\n",
    "    \n",
    "    pattern = f\"[{attr_name[0].lower()}{attr_name[0].upper()}]{attr_name[1:]}=(?P<{attr_name}>[^;]+)\"\n",
    "    match = re.search(pattern, input_str)\n",
    "    if match:\n",
    "        return match.groupdict()[attr_name]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dnaA'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_str = 'ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxref=NCBI_GP:ATV02827.1;Name=ATV02827.1;gbkey=CDS;gene=dnaA;locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "extract_attribute(input_str,\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "input_str = 'locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "assert extract_attribute(input_str,\"gene\") == None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_all_attributes(input_str:str)->dict:\n",
    "    \"\"\"Extracts all attributes from the GFF attributes column\"\"\"\n",
    "    \n",
    "    pattern = \"(?P<key>\\w+[-\\w]*)=(?P<value>[^;]+)\"\n",
    "    match = re.findall(pattern, input_str)\n",
    "    d=defaultdict()\n",
    "    d.update(match)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(None,\n",
       "            {'locus_tag': 'SaO11_00001',\n",
       "             'product': 'Chromosomal replication initiator protein DnaA',\n",
       "             'protein_id': 'ATV02827.1',\n",
       "             'transl_table': '11'})"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_all_attributes(input_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def attributes_to_columns(features: pd.DataFrame):\n",
    "    attr_dicts=features.attributes.apply(extract_all_attributes)\n",
    "    all_keys=list(set().union(*[d.keys() for d in attr_dicts]))\n",
    "    \n",
    "    attr_dict=dict([(k,[d.get(k,None) for d in attr_dicts]) for k in all_keys])\n",
    "    features=features.copy()\n",
    "    for k,v in attr_dict.items():\n",
    "        features[k]=v\n",
    "    \n",
    "    features.fillna(\"\")\n",
    "    return features\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def set_positions(annotation: pd.DataFrame, # an annotation DataFrame extracted from a gff file\n",
    "                            ):\n",
    "    \"\"\"Sets left and right as the position of the feature on the sequence, left is always lower than right.\n",
    "    start and end represent the begining and end of the feature where start can be greater than end depending on the feature strand.\n",
    "    \"\"\"\n",
    "    annotation=annotation.copy()\n",
    "    annotation.loc[:, \"left\"] = annotation[[\"start\"]].values\n",
    "    annotation.loc[:, \"right\"] = annotation[[\"end\"]].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"+\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"left\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"right\"].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"-\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"right\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"left\"].values\n",
    "    \n",
    "    annotation[\"middle\"] = (annotation.right + annotation.left) / 2\n",
    "    \n",
    "    return annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def default_open_gz(gff_path):\n",
    "    if is_gzipped_file(gff_path):\n",
    "        return gzip.open(gff_path,'rt')\n",
    "    else:\n",
    "        return open(gff_path,'r')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "default_types=[\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"]\n",
    "default_attributes=[\"gene\", \"locus_tag\", \"product\"]\n",
    "\n",
    "\n",
    "def parse_gff(gff_path:str, # path to the gff file\n",
    "              seq_id: str=None, # sequence id (first column of the gff)\n",
    "              bounds: tuple=None, # (left limit, right limit)\n",
    "              feature_types: list = None, # list of feature types to extract\n",
    "             )->pd.DataFrame:\n",
    "    cwd = os.getcwd()\n",
    "    \n",
    "    with default_open_gz(gff_path) as gff_file:\n",
    "        # Create an in-memory file buffer using the io.StringIO class\n",
    "        file_buffer = io.StringIO()\n",
    "        default_seq_id=None\n",
    "        buffer_empty=True\n",
    "        for line in gff_file:\n",
    "            if line[0]==\"#\":\n",
    "                continue\n",
    "            else:\n",
    "                r=line.split('\\t')\n",
    "                if not seq_id and not default_seq_id:\n",
    "                    default_seq_id=r[0]\n",
    "                    seq_id=r[0]\n",
    "                if r[0]==seq_id:\n",
    "                    if feature_types==None or r[2] in feature_types:\n",
    "                        if bounds==None or (int(r[3])<bounds[1] and int(r[4])>bounds[0]):\n",
    "                            # Write each line to the file buffer\n",
    "                            file_buffer.write(line)\n",
    "                            buffer_empty=False\n",
    "                        \n",
    "                \n",
    "        # Reset the file pointer to the beginning of the file buffer\n",
    "        file_buffer.seek(0)\n",
    "        if buffer_empty:\n",
    "            warnings.warn(\"The annotation DataFrame is empty. Check that the feature_types and seq_id are correct.\")\n",
    "            df=pd.DataFrame(columns=[\"seq_id\", \"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"])\n",
    "        else:\n",
    "            df=pd.read_csv(file_buffer,sep=\"\\t\",header=None)\n",
    "            df.columns=[\"seq_id\", \"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "            df=attributes_to_columns(df)\n",
    "            df=set_positions(df)\n",
    "     \n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>function</th>\n",
       "      <th>...</th>\n",
       "      <th>ID</th>\n",
       "      <th>pct_identity_gap</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>assembly_bases_seq</th>\n",
       "      <th>regulatory_class</th>\n",
       "      <th>genome</th>\n",
       "      <th>hsp_percent_coverage</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=NC_000001.11:1..248956422;Dbxref=taxon:9606...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_000001.11:1..248956422</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>124478211.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-DDX11L1;Dbxref=GeneID:100287102,HGNC:H...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>gene-DDX11L1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>13141.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_046018.2;Parent=gene-DDX11L1;Dbxref=...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>rna-NR_046018.2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>13141.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-NR_046018.2-1;Parent=rna-NR_046018.2;D...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>exon-NR_046018.2-1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>12050.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-NR_046018.2-2;Parent=rna-NR_046018.2;D...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>exon-NR_046018.2-2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>12667.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seq_id      source        type  start        end score strand phase  \\\n",
       "0  NC_000001.11      RefSeq      region      1  248956422     .      +     .   \n",
       "1  NC_000001.11  BestRefSeq  pseudogene  11874      14409     .      +     .   \n",
       "2  NC_000001.11  BestRefSeq  transcript  11874      14409     .      +     .   \n",
       "3  NC_000001.11  BestRefSeq        exon  11874      12227     .      +     .   \n",
       "4  NC_000001.11  BestRefSeq        exon  12613      12721     .      +     .   \n",
       "\n",
       "                                          attributes function  ...  \\\n",
       "0  ID=NC_000001.11:1..248956422;Dbxref=taxon:9606...     None  ...   \n",
       "1  ID=gene-DDX11L1;Dbxref=GeneID:100287102,HGNC:H...     None  ...   \n",
       "2  ID=rna-NR_046018.2;Parent=gene-DDX11L1;Dbxref=...     None  ...   \n",
       "3  ID=exon-NR_046018.2-1;Parent=rna-NR_046018.2;D...     None  ...   \n",
       "4  ID=exon-NR_046018.2-2;Parent=rna-NR_046018.2;D...     None  ...   \n",
       "\n",
       "                          ID pct_identity_gap chromosome assembly_bases_seq  \\\n",
       "0  NC_000001.11:1..248956422             None          1               None   \n",
       "1               gene-DDX11L1             None       None               None   \n",
       "2            rna-NR_046018.2             None       None               None   \n",
       "3         exon-NR_046018.2-1             None       None               None   \n",
       "4         exon-NR_046018.2-2             None       None               None   \n",
       "\n",
       "  regulatory_class      genome hsp_percent_coverage   left      right  \\\n",
       "0             None  chromosome                 None      1  248956422   \n",
       "1             None        None                 None  11874      14409   \n",
       "2             None        None                 None  11874      14409   \n",
       "3             None        None                 None  11874      12227   \n",
       "4             None        None                 None  12613      12721   \n",
       "\n",
       "        middle  \n",
       "0  124478211.5  \n",
       "1      13141.5  \n",
       "2      13141.5  \n",
       "3      12050.5  \n",
       "4      12667.0  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=parse_gff(human_genome_gff, \n",
    "             seq_id=\"NC_000001.11\",\n",
    "             bounds=(10000,50000))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_feature_types(gff_path):\n",
    "    ftypes=set()\n",
    "    with default_open_gz(gff_path) as handle:\n",
    "        for line in handle:\n",
    "            if line[0]!=\"#\":\n",
    "                r=line.split('\\t')\n",
    "                if len(r)==9:\n",
    "                    ftypes.add(r[2])\n",
    "    return ftypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from genomenotebook.data import get_example_data_dir\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CDS',\n",
       " 'exon',\n",
       " 'gene',\n",
       " 'mobile_genetic_element',\n",
       " 'ncRNA',\n",
       " 'origin_of_replication',\n",
       " 'pseudogene',\n",
       " 'rRNA',\n",
       " 'recombination_feature',\n",
       " 'region',\n",
       " 'repeat_region',\n",
       " 'sequence_feature',\n",
       " 'tRNA'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "available_feature_types(gff_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_attributes(gff_path):\n",
    "    features=parse_gff(gff_path)\n",
    "    return features.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase',\n",
       "       'attributes', 'product', 'locus_tag', 'orig_protein_id', 'pseudo',\n",
       "       'Name', 'gene', 'genome', 'transl_except', 'Dbxref', 'Parent',\n",
       "       'rpt_type', 'gene_biotype', 'orig_transcript_id', 'exception', 'strain',\n",
       "       'gene_synonym', 'part', 'Is_circular', 'Note', 'gbkey', 'substrain',\n",
       "       'mobile_element_type', 'ID', 'transl_table', 'mol_type', 'protein_id',\n",
       "       'recombination_class', 'left', 'right', 'middle'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "available_attributes(gff_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1012/3857466861.py:37: UserWarning: The annotation DataFrame is empty. Check that the feature_types and seq_id are correct.\n",
      "  warnings.warn(\"The annotation DataFrame is empty. Check that the feature_types and seq_id are correct.\")\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "#Testing mistake in seq_id\n",
    "features=parse_gff(gff_path, \"U00097.3\") #mistake in seq_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "gene_y_range = (-1.5, -1)\n",
    "\n",
    "def arrow_coordinates(feature):\n",
    "    feature_size = feature.right - feature.left\n",
    "    \n",
    "    if feature.strand==\"+\":\n",
    "        arrow_base = feature.end - np.minimum(feature_size, 100)\n",
    "    else:\n",
    "        arrow_base = feature.end + np.minimum(feature_size, 100)\n",
    "    \n",
    "    xs=(feature.start,\n",
    "        feature.start,\n",
    "        arrow_base,\n",
    "        feature.end,\n",
    "        arrow_base\n",
    "       )\n",
    "    \n",
    "    y_min, y_max = gene_y_range\n",
    "    ys = (y_min, y_max, y_max, (y_max + y_min) / 2, y_min)\n",
    "    return xs, ys\n",
    "\n",
    "def box_coordinates(feature):\n",
    "    xs=(feature.left, feature.left,\n",
    "        feature.right, feature.right)\n",
    "    y_min, y_max = gene_y_range\n",
    "    ys = (y_min, y_max, y_max, y_min)\n",
    "    return xs, ys\n",
    "\n",
    "default_glyphs=defaultdict(lambda: (\"arrow\",(\"purple\",\"orange\"))) #the default glyph will be the same as for CDS etc.\n",
    "default_glyphs.update(dict([(f,(\"arrow\",(\"purple\",\"orange\"))) for f in [\"CDS\", \"ncRNA\", \"rRNA\", \"tRNA\"]]))\n",
    "default_glyphs['repeat_region']=(\"box\",(\"grey\",))\n",
    "default_glyphs['exon']=(\"box\",(\"grey\",))\n",
    "\n",
    "def get_patch_coordinates(feature, # row of a pandas DataFrame extracted from a GFF file\n",
    "                          patch_dict: dict = default_glyphs # a dictionnary containing as key a feature type and as value a patch definition.\n",
    "                         ):\n",
    "    \"\"\"\n",
    "    patchs are defined with a patch a tuple: (patch_type, (patch_color_plus, patch_color_minus)). \n",
    "    Different colors can be specified depending on the strand.\"\"\"\n",
    "    \n",
    "    patch_type, patch_colors = patch_dict[feature.type]\n",
    "    if len(patch_colors)>1:\n",
    "        color_dic={\"+\":patch_colors[0],\n",
    "                   \"-\":patch_colors[1]}\n",
    "    else:\n",
    "        color_dic=defaultdict(lambda: patch_colors[0])\n",
    "        \n",
    "    if patch_type==\"arrow\":\n",
    "        return arrow_coordinates(feature), color_dic[feature.strand]\n",
    "    elif patch_type==\"box\":\n",
    "        return box_coordinates(feature), color_dic[feature.strand]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((((1, 1, 4641552, 4641652, 4641552), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((190, 190, 190, 255, 190), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((190, 190, 190, 255, 190), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((337, 337, 2699, 2799, 2699), (-1.5, -1, -1, -1.25, -1.5)),\n",
       "  ((337, 337, 2699, 2799, 2699), (-1.5, -1, -1, -1.25, -1.5))),\n",
       " ('purple', 'purple', 'purple', 'purple', 'purple'))"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features=parse_gff(gff_path, \"U00096.3\") #mistake in seq_id\n",
    "coordinates, colors = zip(*features.apply(get_patch_coordinates,axis=1))\n",
    "coordinates[:5], colors[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def split_string(string, max_length=10):\n",
    "    if len(string) <= max_length:\n",
    "        return string\n",
    "    else:\n",
    "        split_index = max_length\n",
    "        while split_index > 0 and string[split_index] != ' ':\n",
    "            split_index -= 1\n",
    "        if split_index == 0:\n",
    "            split_index = max_length  # If no suitable breaking point found, split at max_length\n",
    "        return string[:split_index] + '\\n' + split_string(string[split_index:].lstrip(), max_length)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a very long string that needs to be split\n",
      "into multiple lines because it exceeds 50\n",
      "characters.\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "long_string = \"This is a very long string that needs to be split into multiple lines because it exceeds 50 characters.\"\n",
    "\n",
    "split_result = split_string(long_string, max_length=50)\n",
    "print(split_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_feature_patches(features: pd.DataFrame, #DataFrame of the features \n",
    "                        left: int, #left limit\n",
    "                        right: int, #right limit\n",
    "                        patch_dict: dict = default_glyphs, #glyphs to use for each feature type\n",
    "                        attributes: list = default_attributes, #list of attributes to display when hovering\n",
    "                        name: str = default_attributes[0] #attribute to be displayed as the feature name\n",
    "                       ):\n",
    "    features=features.loc[(features[\"right\"] > left) & (features[\"left\"] < right)]\n",
    "    \n",
    "    if len(features)>0:\n",
    "        coordinates, colors = zip(*features.apply(get_patch_coordinates,patch_dict=patch_dict,axis=1))\n",
    "        xs, ys = zip(*coordinates)\n",
    "    else:\n",
    "        colors = []\n",
    "        xs, ys = [], []\n",
    "        \n",
    "    if name in features.columns:\n",
    "        names = list(features[name].fillna(\"\").values) #list(features.gene.fillna(features[\"locus_tag\"]).values)\n",
    "    else:\n",
    "        names = list(features.iloc[:,9].fillna(\"\").values)\n",
    "        \n",
    "    out=dict(xs=xs,\n",
    "             ys=ys,\n",
    "             color=colors,\n",
    "             pos=features.middle.values,\n",
    "             names=names,\n",
    "             hover_names=names,\n",
    "            )\n",
    "    for attr in attributes:\n",
    "        if attr in features.columns:\n",
    "            values=features[attr].fillna(\"\").astype(str)\n",
    "            out[attr]=values.to_list() #tried to split long strings here but Bokeh then ignores it \n",
    "            \n",
    "    return out\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': ((1, 1, 4641552, 4641652, 4641552),\n",
       "  (10494, 10494, 10028, 9928, 10028),\n",
       "  (10494, 10494, 10028, 9928, 10028),\n",
       "  (11356, 11356, 10743, 10643, 10743),\n",
       "  (11356, 11356, 10743, 10643, 10743),\n",
       "  (10830, 10830, 11215, 11315, 11215),\n",
       "  (10830, 10830, 11215, 11315, 11215)),\n",
       " 'ys': ((-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5)),\n",
       " 'color': ('purple',\n",
       "  'orange',\n",
       "  'orange',\n",
       "  'orange',\n",
       "  'orange',\n",
       "  'purple',\n",
       "  'purple'),\n",
       " 'pos': array([2320826.5,   10211. ,   10211. ,   10999.5,   10999.5,   11072.5,\n",
       "          11072.5]),\n",
       " 'names': ['', 'satP', 'satP', 'yaaW', 'yaaW', 'mbiA', 'mbiA'],\n",
       " 'hover_names': ['', 'satP', 'satP', 'yaaW', 'yaaW', 'mbiA', 'mbiA'],\n",
       " 'gene': ['', 'satP', 'satP', 'yaaW', 'yaaW', 'mbiA', 'mbiA'],\n",
       " 'locus_tag': ['', 'b0010', 'b0010', 'b0011', 'b0011', 'b0012', 'b0012'],\n",
       " 'product': ['',\n",
       "  '',\n",
       "  'acetate/succinate:H(+) symporter',\n",
       "  '',\n",
       "  'putative enzyme-specific chaperone YaaW',\n",
       "  '',\n",
       "  'uncharacterized protein MbiA']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_patches(features,10000,11000, patch_dict=default_glyphs, name=\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "Y_RANGE = (-2, 2)\n",
    "def get_y_range() -> tuple:\n",
    "    \"\"\"Accessor that returns the Y range for the genome browser plot\n",
    "    \"\"\"\n",
    "    return Y_RANGE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def create_genome_browser_plot(glyphSource, x_range, attributes=default_attributes, **kwargs):\n",
    "    plot_height = kwargs.get(\"plot_height\", 150)\n",
    "    label_angle = kwargs.get(\"label_angle\", 45)\n",
    "    text_font_size = kwargs.get(\"text_font_size\", \"10pt\")\n",
    "    output_backend = kwargs.get(\"output_backend\", \"webgl\")\n",
    "    \n",
    "    y_min, y_max = get_y_range()\n",
    "    p_annot = figure(\n",
    "        tools = \"xwheel_zoom,xpan,save\",\n",
    "        active_scroll = \"xwheel_zoom\",\n",
    "        height = plot_height,\n",
    "        x_range = x_range,\n",
    "        y_range = Range1d(y_min, y_max),\n",
    "        output_backend=output_backend,\n",
    "    )\n",
    "    # Add tool\n",
    "    p_annot.add_tools(BoxZoomTool(dimensions=\"width\"))\n",
    "\n",
    "    # Format x axis values\n",
    "    p_annot.xaxis[0].formatter = NumeralTickFormatter(format=\"0,0\")\n",
    "    # Hide grid\n",
    "    p_annot.xgrid.visible = False\n",
    "    p_annot.ygrid.visible = False\n",
    "    # Hide axis\n",
    "    p_annot.yaxis.visible = False\n",
    "    glyph = p_annot.add_glyph(\n",
    "        glyphSource, Patches(xs=\"xs\", ys=\"ys\", fill_color=\"color\")\n",
    "    )\n",
    "    # gene labels in the annotation track\n",
    "    # This seems to be necessary to show the labels\n",
    "    p_annot.scatter(x=\"pos\", y=0, size=0, source=glyphSource)\n",
    "    labels = LabelSet(\n",
    "        x=\"pos\",\n",
    "        y=-0.9,\n",
    "        text=\"names\",\n",
    "        level=\"glyph\",\n",
    "        angle=label_angle,\n",
    "        text_font_size=text_font_size,\n",
    "        x_offset=-5,\n",
    "        y_offset=0,\n",
    "        source=glyphSource,\n",
    "        text_align='left',\n",
    "    )\n",
    "\n",
    "    p_annot.add_layout(labels)\n",
    "    tooltips=[(attr,f\"@{attr}\") for attr in attributes]\n",
    "    p_annot.add_tools(\n",
    "        HoverTool(\n",
    "            renderers=[glyph],\n",
    "            tooltips=tooltips,\n",
    "        )\n",
    "    )\n",
    "    return p_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.models import (\n",
    "    Range1d,\n",
    "    ColumnDataSource,\n",
    ")\n",
    "from bokeh.plotting import show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"25d9361a-8914-4c9c-87dd-b5eac4c91bc3\" data-root-id=\"p1006\" style=\"display: contents;\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "  const docs_json = {\"faf1f15a-7ac5-421c-8847-a41fac414040\":{\"version\":\"3.1.0\",\"title\":\"Bokeh Application\",\"defs\":[],\"roots\":[{\"type\":\"object\",\"name\":\"Figure\",\"id\":\"p1006\",\"attributes\":{\"height\":150,\"x_range\":{\"type\":\"object\",\"name\":\"Range1d\",\"id\":\"p1001\",\"attributes\":{\"start\":2000,\"end\":5000,\"bounds\":[0,5000],\"min_interval\":40,\"max_interval\":100000}},\"y_range\":{\"type\":\"object\",\"name\":\"Range1d\",\"id\":\"p1005\",\"attributes\":{\"start\":-2,\"end\":2}},\"x_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1017\"},\"y_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1019\"},\"title\":{\"type\":\"object\",\"name\":\"Title\",\"id\":\"p1009\"},\"renderers\":[{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1048\",\"attributes\":{\"data_source\":{\"type\":\"object\",\"name\":\"ColumnDataSource\",\"id\":\"p1002\",\"attributes\":{\"selected\":{\"type\":\"object\",\"name\":\"Selection\",\"id\":\"p1004\",\"attributes\":{\"indices\":[],\"line_indices\":[]}},\"selection_policy\":{\"type\":\"object\",\"name\":\"UnionRenderers\",\"id\":\"p1003\"},\"data\":{\"type\":\"map\",\"entries\":[[\"xs\",[[337,337,2699,2799,2699],[2801,2801,3633,3733,3633],[3734,3734,4920,5020,4920]]],[\"ys\",[[-1.5,-1,-1,-1.25,-1.5],[-1.5,-1,-1,-1.25,-1.5],[-1.5,-1,-1,-1.25,-1.5]]],[\"color\",[\"purple\",\"purple\",\"purple\"]],[\"pos\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"AAAAAACAmEAAAAAAAIapQAAAAAAAGbFA\"},\"shape\":[3],\"dtype\":\"float64\",\"order\":\"little\"}],[\"names\",[\"thrA\",\"thrB\",\"thrC\"]],[\"hover_names\",[\"thrA\",\"thrB\",\"thrC\"]],[\"gene\",[\"thrA\",\"thrB\",\"thrC\"]],[\"locus_tag\",[\"b0002\",\"b0003\",\"b0004\"]],[\"product\",[\"fused aspartate kinase/homoserine dehydrogenase 1\",\"homoserine kinase\",\"threonine synthase\"]]]}}},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1049\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1050\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Patches\",\"id\":\"p1047\",\"attributes\":{\"xs\":{\"type\":\"field\",\"field\":\"xs\"},\"ys\":{\"type\":\"field\",\"field\":\"ys\"},\"fill_color\":{\"type\":\"field\",\"field\":\"color\"}}}}},{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1057\",\"attributes\":{\"data_source\":{\"id\":\"p1002\"},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1058\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1059\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1054\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"}}},\"nonselection_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1055\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.1},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.1},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.1}}},\"muted_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1056\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":0},\"size\":{\"type\":\"value\",\"value\":0},\"line_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.2},\"fill_color\":{\"type\":\"value\",\"value\":\"#1f77b4\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.2},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.2}}}}}],\"toolbar\":{\"type\":\"object\",\"name\":\"Toolbar\",\"id\":\"p1012\",\"attributes\":{\"tools\":[{\"type\":\"object\",\"name\":\"WheelZoomTool\",\"id\":\"p1035\",\"attributes\":{\"dimensions\":\"width\"}},{\"type\":\"object\",\"name\":\"PanTool\",\"id\":\"p1036\",\"attributes\":{\"dimensions\":\"width\"}},{\"type\":\"object\",\"name\":\"SaveTool\",\"id\":\"p1037\"},{\"type\":\"object\",\"name\":\"BoxZoomTool\",\"id\":\"p1042\",\"attributes\":{\"dimensions\":\"width\",\"overlay\":{\"type\":\"object\",\"name\":\"BoxAnnotation\",\"id\":\"p1043\",\"attributes\":{\"syncable\":false,\"level\":\"overlay\",\"visible\":false,\"left_units\":\"canvas\",\"right_units\":\"canvas\",\"bottom_units\":\"canvas\",\"top_units\":\"canvas\",\"line_color\":\"black\",\"line_alpha\":1.0,\"line_width\":2,\"line_dash\":[4,4],\"fill_color\":\"lightgrey\",\"fill_alpha\":0.5}}}},{\"type\":\"object\",\"name\":\"HoverTool\",\"id\":\"p1064\",\"attributes\":{\"renderers\":[{\"id\":\"p1048\"}],\"tooltips\":[[\"gene\",\"@gene\"],[\"locus_tag\",\"@locus_tag\"],[\"product\",\"@product\"]]}}],\"active_scroll\":{\"id\":\"p1035\"}}},\"left\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1028\",\"attributes\":{\"visible\":false,\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1029\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1030\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1031\"}}}],\"below\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1021\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1022\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"NumeralTickFormatter\",\"id\":\"p1045\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1024\"}}}],\"center\":[{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1027\",\"attributes\":{\"visible\":false,\"axis\":{\"id\":\"p1021\"}}},{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1034\",\"attributes\":{\"visible\":false,\"dimension\":1,\"axis\":{\"id\":\"p1028\"}}},{\"type\":\"object\",\"name\":\"LabelSet\",\"id\":\"p1060\",\"attributes\":{\"level\":\"glyph\",\"source\":{\"id\":\"p1002\"},\"x\":{\"type\":\"field\",\"field\":\"pos\"},\"y\":{\"type\":\"value\",\"value\":-0.9},\"text\":{\"type\":\"field\",\"field\":\"names\"},\"angle\":{\"type\":\"value\",\"value\":45},\"x_offset\":{\"type\":\"value\",\"value\":-5},\"text_font_size\":{\"type\":\"value\",\"value\":\"10pt\"}}}],\"output_backend\":\"webgl\"}}],\"callbacks\":{\"type\":\"map\"}}};\n",
       "  const render_items = [{\"docid\":\"faf1f15a-7ac5-421c-8847-a41fac414040\",\"roots\":{\"p1006\":\"25d9361a-8914-4c9c-87dd-b5eac4c91bc3\"},\"root_ids\":[\"p1006\"]}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    let attempts = 0;\n",
       "    const timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "p1006"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "features = parse_gff(gff_path,\n",
    "                     feature_types=default_types,\n",
    "                    )\n",
    "x_range = Range1d(\n",
    "            2000, 5000, \n",
    "            bounds=(0,5000), \n",
    "            max_interval=100000,\n",
    "            min_interval=40\n",
    "        )\n",
    "\n",
    "feature_patches = get_feature_patches(features, x_range.start, x_range.end)\n",
    "glyph_source = ColumnDataSource(feature_patches)\n",
    "p = create_genome_browser_plot(glyph_source, x_range)\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': ((337, 337, 2699, 2799, 2699),\n",
       "  (2801, 2801, 3633, 3733, 3633),\n",
       "  (3734, 3734, 4920, 5020, 4920)),\n",
       " 'ys': ((-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5),\n",
       "  (-1.5, -1, -1, -1.25, -1.5)),\n",
       " 'color': ('purple', 'purple', 'purple'),\n",
       " 'pos': array([1568., 3267., 4377.]),\n",
       " 'names': ['thrA', 'thrB', 'thrC'],\n",
       " 'hover_names': ['thrA', 'thrB', 'thrC'],\n",
       " 'gene': ['thrA', 'thrB', 'thrC'],\n",
       " 'locus_tag': ['b0002', 'b0003', 'b0004'],\n",
       " 'product': ['fused aspartate kinase/homoserine dehydrogenase 1',\n",
       "  'homoserine kinase',\n",
       "  'threonine synthase']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_patches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "def sort_list_dict(d: dict, #a dictionnary for which all values are of type list\n",
    "                   ref_list=\"xs\", #key of the list to use as a reference for sorting\n",
    "                   func=lambda x: x, #A custom function can be supplied to customize the sort order. Default is the identity function.\n",
    "                  ):\n",
    "    ks=list(d.keys())\n",
    "    ref_list_ix=ks.index(ref_list)\n",
    "    # Sort all the lists in the dictionary based on the values of the reference list\n",
    "    sorted_lists = sorted(zip(*[d[k] for k in ks]), key= lambda x: func(x[ref_list_ix]))\n",
    "\n",
    "    # Convert the sorted tuples back into separate lists\n",
    "    unzipped_lists = zip(*sorted_lists)\n",
    "\n",
    "    # Create a new dictionary with the same keys as the original dictionary, but with the sorted lists as values\n",
    "    d = {k: list(t) for k, t in zip(ks, unzipped_lists)}\n",
    "    return d\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': [[3, 1], [2, 5]], 'a': [3, 1], 'b': ['d', 'c']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "d={'xs':[[2,5],[3,1]],'a':[1,3],'b':[\"c\",\"d\"]}\n",
    "sort_list_dict(d, ref_list='xs', func= lambda x: x[1]) #sort according to the second element of each element of list 'xs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
