{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# utils\n",
    "\n",
    "> This contains useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.2.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from bokeh.io import output_notebook #|hide_line\n",
    "output_notebook(hide_banner=True) #|hide_line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import io\n",
    "\n",
    "from collections import defaultdict, OrderedDict\n",
    "import warnings\n",
    "import gzip\n",
    "import urllib.request\n",
    "import os\n",
    "import re\n",
    "from platform import uname\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "from typing import List, Optional, Dict, Tuple\n",
    "from IPython.display import display, HTML\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def download_file(url, save_path):\n",
    "    \"\"\"Checks if a file with the same name is already in the save_path. If not download it.\"\"\"\n",
    "    if os.path.exists(save_path):\n",
    "        print(f\"File already exists: {save_path}\")\n",
    "    else:\n",
    "        urllib.request.urlretrieve(url, save_path)\n",
    "        print(f\"File downloaded and saved: {save_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def is_gzipped_file(file_path):\n",
    "    try:\n",
    "        with gzip.open(file_path, 'rb') as f:\n",
    "            # Attempt to read a small chunk from the file\n",
    "            f.read(1)\n",
    "        return True\n",
    "    except IOError:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def default_open_gz(gff_path):\n",
    "    \"\"\"If file is gzipped then opens it with `gzip.open`, otherwise opens it with `open`\"\"\"\n",
    "    if is_gzipped_file(gff_path):\n",
    "        return gzip.open(gff_path,'rt')\n",
    "    else:\n",
    "        return open(gff_path,'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists: GRCh38_latest_genomic.gff.gz\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example usage\n",
    "file_url = 'https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gff.gz'\n",
    "human_genome_gff = 'GRCh38_latest_genomic.gff.gz'\n",
    "\n",
    "download_file(file_url, human_genome_gff)\n",
    "is_gzipped_file(human_genome_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_attribute(input_str:str, #attribute string to parse\n",
    "                      attr_name:str, #name of the attribute to extract\n",
    "                     ) -> str:\n",
    "    \"\"\"Extracts the attribute called attr_name from the GFF attributes string\"\"\"\n",
    "    \n",
    "    pattern = f\"[{attr_name[0].lower()}{attr_name[0].upper()}]{attr_name[1:]}=(?P<{attr_name}>[^;]+)\"\n",
    "    match = re.search(pattern, input_str)\n",
    "    if match:\n",
    "        return match.groupdict()[attr_name]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dnaA'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_str = 'ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxref=NCBI_GP:ATV02827.1;Name=ATV02827.1;gbkey=CDS;gene=dnaA;locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "extract_attribute(input_str,\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "input_str = 'locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'\n",
    "assert extract_attribute(input_str,\"gene\") == None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_all_attributes(input_str:str)->OrderedDict: #TODO: why is this not limited by the attributes subset provided to GenomeBrowser?\n",
    "    \"\"\"Extracts all attributes from the GFF attributes column\"\"\"\n",
    "    \n",
    "    pattern = \"(?P<key>\\w+[-\\w]*)=(?P<value>[^;]+)\"\n",
    "    match = re.findall(pattern, input_str)\n",
    "    d=OrderedDict()\n",
    "    d.update(match)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def extract_attributes(input_str:str, #the attribute string of a GFF fome\n",
    "                       attributes: Optional[List[str]] = None #an optional list of attribute names to extract. If None all attributes are extracted.\n",
    "                       )->OrderedDict: \n",
    "    \"\"\"Extracts attributes from the GFF attributes column\"\"\"\n",
    "    pattern = \"(?P<key>\\w+[-\\w]*)=(?P<value>[^;]+)\"\n",
    "    match = re.findall(pattern, input_str)\n",
    "    d=OrderedDict()\n",
    "    if attributes is not None:\n",
    "        match = [m for m in match if m[0] in attributes]\n",
    "    d.update(match)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_attributes(df: pd.DataFrame, #a features DataFrame with at least a \"type\" column and an \"attributes_str\" column\n",
    "                   attributes: Optional[Dict[str, List]] = None # a dictionary with feature types as keys and a list of attributes to extract as values \n",
    "                   ) -> List:\n",
    "    \"\"\"Iterates over each row of the df and extracts the attributes specified in the attributes dictionary for each feature type\"\"\"\n",
    "    attr_list=[]\n",
    "    for i, row in df.iterrows():\n",
    "        if attributes is None:\n",
    "            attrs = None\n",
    "        elif row.type in attributes:\n",
    "            attrs = attributes[row.type]\n",
    "        else:\n",
    "            attrs = None\n",
    "        \n",
    "        attr_list.append(extract_attributes(row.attributes_str,attrs))\n",
    "\n",
    "    return attr_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def attributes_to_columns(features: pd.DataFrame):\n",
    "    attr_dicts=features.attributes.apply(extract_all_attributes)\n",
    "    all_keys=list(set().union(*[d.keys() for d in attr_dicts]))\n",
    "    \n",
    "    attr_dict=dict([(k,[d.get(k,None) for d in attr_dicts]) for k in all_keys])\n",
    "    features=features.copy()\n",
    "    for k,v in attr_dict.items():\n",
    "        features[k]=v\n",
    "    \n",
    "    features.fillna(\"\")\n",
    "    return features\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def set_positions(annotation: pd.DataFrame, # an annotation DataFrame extracted from a gff file\n",
    "                            ) ->  pd.DataFrame:\n",
    "    \"\"\"Sets left and right as the position of the feature on the sequence, left is always lower than right.\n",
    "    start and end represent the begining and end of the feature where start can be greater than end depending on the feature strand.\n",
    "    \"\"\"\n",
    "    annotation=annotation.copy()\n",
    "    annotation.loc[:, \"left\"] = annotation[[\"start\"]].values\n",
    "    annotation.loc[:, \"right\"] = annotation[[\"end\"]].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"+\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"left\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"right\"].values\n",
    "    \n",
    "    mask = annotation[\"strand\"] == \"-\"\n",
    "    annotation.loc[mask, \"start\"] = annotation.loc[mask, \"right\"].values\n",
    "    annotation.loc[mask, \"end\"] = annotation.loc[mask, \"left\"].values\n",
    "    \n",
    "    annotation[\"middle\"] = (annotation.right + annotation.left) / 2\n",
    "    \n",
    "    return annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#| hide\n",
    "class EmptyDataFrame(Exception):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def parse_gff(gff_path:str, # path to the gff file\n",
    "              seq_id: Optional[str] = None, # sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name\n",
    "              first: bool = True, # if True then return only the annotations for the first sequence (or the first with seq_id)\n",
    "              bounds: Optional[tuple] = None, # (left limit, right limit)\n",
    "              feature_types: Optional[list] = None, # list of feature types to extract\n",
    "              attributes: Optional[Dict[str, List]] = None, # a dictionary with feature types as keys and a list of attributes to extract as values \n",
    "             )->List[pd.DataFrame]:\n",
    "    \"\"\" Parses a GFF3 file and returns a list of Pandas DataFrames with the data for a specific contig. \n",
    "    If seq_id is None then only the first contig is parsed.\n",
    "    If feature_types is None then all feature types are extracted.\"\"\"\n",
    "\n",
    "    if attributes is None:\n",
    "        attributes = {}\n",
    "\n",
    "    def _slurp_buffer(file_buffer, buffer_empty):\n",
    "        # Reset the file pointer to the beginning of the file buffer\n",
    "        file_buffer.seek(0)\n",
    "        if buffer_empty:\n",
    "            raise EmptyDataFrame(\"The annotation DataFrame is empty. Check that the feature_types and seq_id are correct, and that bounds (if specified) fall within the size of your genome.\")\n",
    "            df=pd.DataFrame(columns=[\"seq_id\", \"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"])\n",
    "        else:\n",
    "            df=pd.read_csv(file_buffer,sep=\"\\t\",header=None)\n",
    "            df.columns=[\"seq_id\", \"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes_str\"]\n",
    "            #df=attributes_to_columns(df)\n",
    "            df[\"attributes\"] = get_attributes(df, attributes)\n",
    "            df.drop(columns=[\"attributes_str\"], inplace=True)\n",
    "            df=set_positions(df)\n",
    "        return df\n",
    "    \n",
    "    out = list()\n",
    "\n",
    "    #NOTE: This assumes that all lines for a given seq_id are consecutive, which is generally the case for gff files.\n",
    "    with default_open_gz(gff_path) as gff_file:\n",
    "        # Create an in-memory file buffer using the io.StringIO class\n",
    "        file_buffer = io.StringIO()\n",
    "        buffer_empty = True\n",
    "        last_seq_id = None\n",
    "        for line in gff_file:\n",
    "            if line[0]==\"#\":\n",
    "                continue\n",
    "            else:\n",
    "                r=line.split('\\t')\n",
    "                current_line_seqid = r[0]\n",
    "                if last_seq_id is not None and current_line_seqid != last_seq_id: #seeing a new segment of the gff\n",
    "                    if not buffer_empty:\n",
    "                        if seq_id is None or seq_id == last_seq_id:\n",
    "                            out.append(_slurp_buffer(file_buffer, buffer_empty))\n",
    "                            file_buffer = io.StringIO()\n",
    "                            buffer_empty = True\n",
    "                            if first is not None or seq_id is not None:\n",
    "                                break\n",
    "                last_seq_id = current_line_seqid\n",
    "                if seq_id is None: #\n",
    "                    seq_id = current_line_seqid\n",
    "                if r[0]==seq_id:\n",
    "                    if feature_types==None or r[2] in feature_types:\n",
    "                        if bounds==None or (int(r[3])<bounds[1] and int(r[4])>bounds[0]):\n",
    "                            # Write each line to the file buffer\n",
    "                            file_buffer.write(line)\n",
    "                            buffer_empty=False\n",
    "        if not buffer_empty:\n",
    "            if seq_id is None or seq_id == last_seq_id:\n",
    "                out.append(_slurp_buffer(file_buffer, buffer_empty))\n",
    "    \n",
    "    if len(out) == 0:\n",
    "        raise EmptyDataFrame(\"The annotation DataFrame is empty. Check that the feature_types and seq_id are correct, and that bounds (if specified) fall within the size of your genome.\")\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from genomenotebook.data import get_example_data_dir\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'NZ_JAGURL010000100.1:1..16949', 'Dbxre...</td>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>8475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10059</td>\n",
       "      <td>8311</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS20985', 'Name': 'KFX61_RS...</td>\n",
       "      <td>8311</td>\n",
       "      <td>10059</td>\n",
       "      <td>9185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10059</td>\n",
       "      <td>8311</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_225638104.1', 'Parent': 'gene-K...</td>\n",
       "      <td>8311</td>\n",
       "      <td>10059</td>\n",
       "      <td>9185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10541</td>\n",
       "      <td>10092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS20990', 'Name': 'KFX61_RS...</td>\n",
       "      <td>10092</td>\n",
       "      <td>10541</td>\n",
       "      <td>10316.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10541</td>\n",
       "      <td>10092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_016268480.1', 'Parent': 'gene-K...</td>\n",
       "      <td>10092</td>\n",
       "      <td>10541</td>\n",
       "      <td>10316.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seq_id            source    type  start    end score strand  \\\n",
       "0  NZ_JAGURL010000100.1            RefSeq  region      1  16949     .      +   \n",
       "1  NZ_JAGURL010000100.1            RefSeq    gene  10059   8311     .      -   \n",
       "2  NZ_JAGURL010000100.1  Protein Homology     CDS  10059   8311     .      -   \n",
       "3  NZ_JAGURL010000100.1            RefSeq    gene  10541  10092     .      -   \n",
       "4  NZ_JAGURL010000100.1  Protein Homology     CDS  10541  10092     .      -   \n",
       "\n",
       "  phase                                         attributes   left  right  \\\n",
       "0     .  {'ID': 'NZ_JAGURL010000100.1:1..16949', 'Dbxre...      1  16949   \n",
       "1     .  {'ID': 'gene-KFX61_RS20985', 'Name': 'KFX61_RS...   8311  10059   \n",
       "2     0  {'ID': 'cds-WP_225638104.1', 'Parent': 'gene-K...   8311  10059   \n",
       "3     .  {'ID': 'gene-KFX61_RS20990', 'Name': 'KFX61_RS...  10092  10541   \n",
       "4     0  {'ID': 'cds-WP_016268480.1', 'Parent': 'gene-K...  10092  10541   \n",
       "\n",
       "    middle  \n",
       "0   8475.0  \n",
       "1   9185.0  \n",
       "2   9185.0  \n",
       "3  10316.5  \n",
       "4  10316.5  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"jmh43.gff\")\n",
    "df=parse_gff(gff_path, \n",
    "             bounds=(10000,50000))[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>98838</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'NZ_JAGURL010000013.1:1..98838', 'Dbxre...</td>\n",
       "      <td>1</td>\n",
       "      <td>98838</td>\n",
       "      <td>49419.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10195</td>\n",
       "      <td>11175</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS05880', 'Name': 'KFX61_RS...</td>\n",
       "      <td>10195</td>\n",
       "      <td>11175</td>\n",
       "      <td>10685.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10195</td>\n",
       "      <td>11175</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_048697200.1', 'Parent': 'gene-K...</td>\n",
       "      <td>10195</td>\n",
       "      <td>11175</td>\n",
       "      <td>10685.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>11234</td>\n",
       "      <td>11824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS05885', 'Name': 'KFX61_RS...</td>\n",
       "      <td>11234</td>\n",
       "      <td>11824</td>\n",
       "      <td>11529.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>11234</td>\n",
       "      <td>11824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_048697198.1', 'Parent': 'gene-K...</td>\n",
       "      <td>11234</td>\n",
       "      <td>11824</td>\n",
       "      <td>11529.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seq_id            source    type  start    end score strand  \\\n",
       "0  NZ_JAGURL010000013.1            RefSeq  region      1  98838     .      +   \n",
       "1  NZ_JAGURL010000013.1            RefSeq    gene  10195  11175     .      +   \n",
       "2  NZ_JAGURL010000013.1  Protein Homology     CDS  10195  11175     .      +   \n",
       "3  NZ_JAGURL010000013.1            RefSeq    gene  11234  11824     .      +   \n",
       "4  NZ_JAGURL010000013.1  Protein Homology     CDS  11234  11824     .      +   \n",
       "\n",
       "  phase                                         attributes   left  right  \\\n",
       "0     .  {'ID': 'NZ_JAGURL010000013.1:1..98838', 'Dbxre...      1  98838   \n",
       "1     .  {'ID': 'gene-KFX61_RS05880', 'Name': 'KFX61_RS...  10195  11175   \n",
       "2     0  {'ID': 'cds-WP_048697200.1', 'Parent': 'gene-K...  10195  11175   \n",
       "3     .  {'ID': 'gene-KFX61_RS05885', 'Name': 'KFX61_RS...  11234  11824   \n",
       "4     0  {'ID': 'cds-WP_048697198.1', 'Parent': 'gene-K...  11234  11824   \n",
       "\n",
       "    middle  \n",
       "0  49419.5  \n",
       "1  10685.0  \n",
       "2  10685.0  \n",
       "3  11529.0  \n",
       "4  11529.0  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=parse_gff(gff_path, \n",
    "             seq_id=\"NZ_JAGURL010000013.1\",\n",
    "             bounds=(10000,50000))[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#testing feature_types\n",
    "gff_path = os.path.join(data_path, \"jmh43.gff\")\n",
    "df=parse_gff(gff_path, \n",
    "             seq_id=\"NZ_JAGURL010000013.1\",\n",
    "             feature_types=[\"gene\"],\n",
    "             bounds=(10000,50000))[0]\n",
    "assert(df.loc[0,\"type\"]==\"gene\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'NZ_JAGURL010000100.1:1..16949', 'Dbxre...</td>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>8475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10059</td>\n",
       "      <td>8311</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS20985', 'Name': 'KFX61_RS...</td>\n",
       "      <td>8311</td>\n",
       "      <td>10059</td>\n",
       "      <td>9185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10059</td>\n",
       "      <td>8311</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_225638104.1'}</td>\n",
       "      <td>8311</td>\n",
       "      <td>10059</td>\n",
       "      <td>9185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10541</td>\n",
       "      <td>10092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS20990', 'Name': 'KFX61_RS...</td>\n",
       "      <td>10092</td>\n",
       "      <td>10541</td>\n",
       "      <td>10316.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_JAGURL010000100.1</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10541</td>\n",
       "      <td>10092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-WP_016268480.1'}</td>\n",
       "      <td>10092</td>\n",
       "      <td>10541</td>\n",
       "      <td>10316.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seq_id            source    type  start    end score strand  \\\n",
       "0  NZ_JAGURL010000100.1            RefSeq  region      1  16949     .      +   \n",
       "1  NZ_JAGURL010000100.1            RefSeq    gene  10059   8311     .      -   \n",
       "2  NZ_JAGURL010000100.1  Protein Homology     CDS  10059   8311     .      -   \n",
       "3  NZ_JAGURL010000100.1            RefSeq    gene  10541  10092     .      -   \n",
       "4  NZ_JAGURL010000100.1  Protein Homology     CDS  10541  10092     .      -   \n",
       "\n",
       "  phase                                         attributes   left  right  \\\n",
       "0     .  {'ID': 'NZ_JAGURL010000100.1:1..16949', 'Dbxre...      1  16949   \n",
       "1     .  {'ID': 'gene-KFX61_RS20985', 'Name': 'KFX61_RS...   8311  10059   \n",
       "2     0                       {'ID': 'cds-WP_225638104.1'}   8311  10059   \n",
       "3     .  {'ID': 'gene-KFX61_RS20990', 'Name': 'KFX61_RS...  10092  10541   \n",
       "4     0                       {'ID': 'cds-WP_016268480.1'}  10092  10541   \n",
       "\n",
       "    middle  \n",
       "0   8475.0  \n",
       "1   9185.0  \n",
       "2   9185.0  \n",
       "3  10316.5  \n",
       "4  10316.5  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "#testing attributes\n",
    "attributes = {\"gene\":[\"Name\",\"ID\",\"X\"], \"CDS\":[\"ID\"]}\n",
    "df=parse_gff(gff_path, \n",
    "             attributes=attributes,\n",
    "             bounds=(10000,50000))[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#testing EmptyDataFrame Error\n",
    "try:\n",
    "    df=parse_gff(gff_path, \n",
    "                seq_id=\"NZ_JAGURL010000013.1\",\n",
    "                feature_types=[\"x\"],\n",
    "                bounds=(10000,50000))[0]\n",
    "except EmptyDataFrame:\n",
    "    exception_raised = True\n",
    "\n",
    "assert exception_raised\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#Testing mistake in seq_id\n",
    "try:\n",
    "    features=parse_gff(gff_path, \"U00097.3\") #mistake in seq_id\n",
    "except EmptyDataFrame:\n",
    "    exception_raised = True\n",
    "\n",
    "assert exception_raised"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'U00096.3:1..4641652', 'Dbxref': 'taxon...</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>2320826.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-b0001', 'Dbxref': 'ASAP:ABE-00000...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-AAC73112.1', 'Parent': 'gene-b0001...</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>222.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-b0002', 'Dbxref': 'ASAP:ABE-00000...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>{'ID': 'cds-AAC73113.1', 'Parent': 'gene-b0002...</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>1568.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id   source    type  start      end score strand phase  \\\n",
       "0  U00096.3  Genbank  region      1  4641652     .      +     .   \n",
       "1  U00096.3  Genbank    gene    190      255     .      +     .   \n",
       "2  U00096.3  Genbank     CDS    190      255     .      +     0   \n",
       "3  U00096.3  Genbank    gene    337     2799     .      +     .   \n",
       "4  U00096.3  Genbank     CDS    337     2799     .      +     0   \n",
       "\n",
       "                                          attributes  left    right     middle  \n",
       "0  {'ID': 'U00096.3:1..4641652', 'Dbxref': 'taxon...     1  4641652  2320826.5  \n",
       "1  {'ID': 'gene-b0001', 'Dbxref': 'ASAP:ABE-00000...   190      255      222.5  \n",
       "2  {'ID': 'cds-AAC73112.1', 'Parent': 'gene-b0001...   190      255      222.5  \n",
       "3  {'ID': 'gene-b0002', 'Dbxref': 'ASAP:ABE-00000...   337     2799     1568.0  \n",
       "4  {'ID': 'cds-AAC73113.1', 'Parent': 'gene-b0002...   337     2799     1568.0  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_path = os.path.join(data_path, \"MG1655_U00096.fasta\")\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "parse_gff(gff_path)[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_feature_types(gff_path):\n",
    "    ftypes=set()\n",
    "    with default_open_gz(gff_path) as handle:\n",
    "        for line in handle:\n",
    "            if line[0]!=\"#\":\n",
    "                r=line.split('\\t')\n",
    "                if len(r)==9:\n",
    "                    ftypes.add(r[2])\n",
    "    return ftypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CDS',\n",
       " 'exon',\n",
       " 'gene',\n",
       " 'mobile_genetic_element',\n",
       " 'ncRNA',\n",
       " 'origin_of_replication',\n",
       " 'pseudogene',\n",
       " 'rRNA',\n",
       " 'recombination_feature',\n",
       " 'region',\n",
       " 'repeat_region',\n",
       " 'sequence_feature',\n",
       " 'tRNA'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = get_example_data_dir()\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "available_feature_types(gff_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def available_attributes(gff_path):\n",
    "    features=parse_gff(gff_path)[0]\n",
    "    return features.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase',\n",
       "       'attributes', 'left', 'right', 'middle'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "available_attributes(gff_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def parse_fasta(genome_path, seq_id):\n",
    "    \"\"\"Retrieves the Biopython SeqRecord object that matches the seq_id in a fasta file\"\"\"\n",
    "\n",
    "    rec_found=False\n",
    "    with open(genome_path,'r') as f:\n",
    "        for rec in SeqIO.parse(f, 'fasta'):\n",
    "            if rec.id==seq_id:\n",
    "                rec_found=True\n",
    "                break\n",
    "\n",
    "    if not rec_found:\n",
    "        warnings.warn(\"seq_id not found in fasta file\")\n",
    "        rec = None\n",
    "    \n",
    "    return rec.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "faa_path = os.path.join(data_path, \"MG1655_U00096.fasta\")\n",
    "rec = next(SeqIO.parse(faa_path, 'fasta'))\n",
    "testseq = str(rec.seq)\n",
    "\n",
    "rec = parse_fasta(faa_path,rec.id)\n",
    "assert(str(rec) == testseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def regions_overlap(region1, region2, min_overlap_fraction=0.0):\n",
    "    \"\"\"\n",
    "        regions are tuples of start and stop coordinates\n",
    "        returns true if a fraction of region2 >= min_overlap_fraction overlaps with region1\n",
    "        coordinates within regions must be sorted low to high\n",
    "    \"\"\"\n",
    "\n",
    "    if min_overlap_fraction >= 1:\n",
    "        return False\n",
    "    \n",
    "    # incoming coordinates are gff-base (1-based, inclusive), convert to python-base (0-based)\n",
    "\n",
    "    region_1 = (region1[0] - 1, region1[1])\n",
    "    region_2 = (region2[0] - 1, region2[1])\n",
    "    region_1_size = region1[1] - region1[0]\n",
    "    region_2_size = region2[1] - region2[0]\n",
    "    if region_1_size == 0:\n",
    "        region_1_size = 0.1\n",
    "    if region_2_size == 0:\n",
    "        region_2_size = 0.1\n",
    "\n",
    "    # -region2 start is contained in region11\n",
    "    if region2[0] <= region1[1] and region2[0] >= region1[0]:\n",
    "        # not true for completely contained, but we don't care, should technically be min(region2[1] - region2[0], region1[1] - region2[0])\n",
    "        overlap_size = region1[1] - region2[0]\n",
    "        if (overlap_size / region_2_size) >= min_overlap_fraction:\n",
    "            return True\n",
    "\n",
    "    # -region2 end is contained in region1\n",
    "    if region2[1] <= region1[1] and region2[1] >= region1[0]:\n",
    "        # not true for completely contained, but we don't care\n",
    "        overlap_size = region2[1] - region1[0]\n",
    "        if (overlap_size / region_2_size) >= min_overlap_fraction:\n",
    "            return True\n",
    "\n",
    "    # -region1 start is contained in region2\n",
    "    if region1[0] <= region2[1] and region1[0] >= region2[0]:\n",
    "        # not true for completely contained, but we don't care\n",
    "        overlap_size = region2[1] - region1[0]\n",
    "        if (overlap_size / region_2_size) >= min_overlap_fraction:\n",
    "            return True\n",
    "\n",
    "    # -region1 end is contained in region2\n",
    "    if region1[1] <= region2[1] and region1[1] >= region2[0]:\n",
    "        # not true for completely contained, but we don't care\n",
    "        overlap_size = region1[1] - region2[0]\n",
    "        if (overlap_size / region_2_size) >= min_overlap_fraction:\n",
    "            return True\n",
    "\n",
    "    return False\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "assert(regions_overlap((0,100),(90,200))==True)\n",
    "assert(regions_overlap((0,100),(190,200))==False)\n",
    "assert(regions_overlap((190,200),(0,100))==False)\n",
    "assert(regions_overlap((190,200),(0,195))==True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def add_z_order(features, \n",
    "                prescedence = [\"source\", \"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\",\"exon\"]):\n",
    "    \"\"\"\n",
    "        features is a dataframe of features\n",
    "        prescedence is a list of feature types in order of prescedence, e.g. [\"CDS\", \"repeat_region\", \"ncRNA\", \"rRNA\", \"tRNA\"] will put \"CDS\" features closer to the bottom of the plot than \"repeat_region\" features.\n",
    "        returns features with a z_order column added\n",
    "    \"\"\"\n",
    "    #TODO: possibility for \"linking attributes\" to link features and cause them to have the same z-order and occupy their entire envelope.\n",
    "    type_order = defaultdict(lambda: len(prescedence)+1)\n",
    "    type_order.update({t: i for i, t in enumerate(prescedence)})\n",
    "    features.sort_values(by=\"start\", inplace=True)\n",
    "    features.sort_values(by=\"type\", inplace=True, key=lambda x: x.map(type_order))\n",
    "    z_order = []\n",
    "    added = []\n",
    "    all_z = {0}\n",
    "    for index, row in features.iterrows():\n",
    "        left, right = row[\"left\"], row[\"right\"]\n",
    "        z = 0\n",
    "        z_found = set()\n",
    "        for (l_a, r_a, z_a, z_o) in added:\n",
    "            if regions_overlap((left, right), (l_a, r_a)):\n",
    "                if type_order[row[\"type\"]] > z_o:\n",
    "                    for i in range(z_a+1):\n",
    "                        z_found.add(i)\n",
    "                else:\n",
    "                    z_found.add(z_a)\n",
    "        if len(z_found) == len(all_z):\n",
    "            z = max(all_z) + 1\n",
    "            all_z.add(z)\n",
    "        else:\n",
    "            z = min(all_z - z_found)\n",
    "        z_order.append(z)\n",
    "        added.append((left, right, z, type_order[row[\"type\"]]))\n",
    "    features[\"z_order\"] = z_order\n",
    "\n",
    "    features.sort_values(by=\"start\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>middle</th>\n",
       "      <th>z_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>14538</td>\n",
       "      <td>16334</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS05895', 'Name': 'KFX61_RS...</td>\n",
       "      <td>14538</td>\n",
       "      <td>16334</td>\n",
       "      <td>15436.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16377</td>\n",
       "      <td>16847</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS05900', 'Name': 'KFX61_RS...</td>\n",
       "      <td>16377</td>\n",
       "      <td>16847</td>\n",
       "      <td>16612.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NZ_JAGURL010000013.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>18894</td>\n",
       "      <td>16822</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'ID': 'gene-KFX61_RS05905', 'Name': 'KFX61_RS...</td>\n",
       "      <td>16822</td>\n",
       "      <td>18894</td>\n",
       "      <td>17858.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seq_id  source  type  start    end score strand phase  \\\n",
       "3  NZ_JAGURL010000013.1  RefSeq  gene  14538  16334     .      +     .   \n",
       "4  NZ_JAGURL010000013.1  RefSeq  gene  16377  16847     .      +     .   \n",
       "5  NZ_JAGURL010000013.1  RefSeq  gene  18894  16822     .      -     .   \n",
       "\n",
       "                                          attributes   left  right   middle  \\\n",
       "3  {'ID': 'gene-KFX61_RS05895', 'Name': 'KFX61_RS...  14538  16334  15436.0   \n",
       "4  {'ID': 'gene-KFX61_RS05900', 'Name': 'KFX61_RS...  16377  16847  16612.0   \n",
       "5  {'ID': 'gene-KFX61_RS05905', 'Name': 'KFX61_RS...  16822  18894  17858.0   \n",
       "\n",
       "   z_order  \n",
       "3        0  \n",
       "4        0  \n",
       "5        1  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "gff_path = os.path.join(data_path, \"jmh43.gff\")\n",
    "df=parse_gff(gff_path, \n",
    "             seq_id=\"NZ_JAGURL010000013.1\",\n",
    "             feature_types=[\"gene\"],\n",
    "             bounds=(10000,50000))[0]\n",
    "add_z_order(df)\n",
    "df.iloc[3:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "#### Code from Domainator\n",
    "def get_cds_unique_name(feature):\n",
    "    \"\"\"\n",
    "        If the feature already has a cds_id, then keep it, otherwise generate one based on the position on the contig.\n",
    "    \"\"\"\n",
    "    if \"cds_id\" in feature.qualifiers:\n",
    "        return feature.qualifiers[\"cds_id\"][0]\n",
    "    else:\n",
    "        # need the strand information to account for circular contigs.\n",
    "        name_parts = [\"_\".join( (str(p.stranded_start_human_readable), str(p.strand), str(p.stranded_end_human_readable)) ) for p in feature.location.parts]\n",
    "        return \" \".join(name_parts) # space so it can be split into multiple lines when writing genbank files\n",
    "\n",
    "def get_cds_name(feature): #(contig_id, feature):\n",
    "    if \"gene_id\" in feature.qualifiers:\n",
    "        return feature.qualifiers[\"gene_id\"][0]\n",
    "    elif \"locus_tag\" in feature.qualifiers:\n",
    "        return feature.qualifiers[\"locus_tag\"][0]\n",
    "    else:\n",
    "        return get_cds_unique_name(feature)\n",
    "#### End code from Domainator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from Bio import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "strand_dict = {1: \"+\", -1: \"-\"}\n",
    "\n",
    "def seqRecord_to_df(rec: SeqRecord,\n",
    "                    feature_types: Optional[List[str]] = None, # if None then get all features, otherwise only those with type in FeatureTypes.\n",
    "                    attributes: Optional[Dict[str,List]] = None \n",
    "                    # if None, then get all attributes of all feature types. If dict, then only get attributes of feature types keys. If value is None, get all\n",
    "                    )->pd.DataFrame:\n",
    "                    \n",
    "    feature_lists = []\n",
    "    for feature in rec.features:\n",
    "        if feature_types is None or feature.type in feature_types:\n",
    "            if attributes is None:\n",
    "                attrs = None\n",
    "            else:\n",
    "                attrs=attributes.get(feature.type, None)\n",
    "            for part in feature.location.parts:\n",
    "                attributes_list = []#[(\"ID\", get_cds_name(feature)),]\n",
    "                for key, value in feature.qualifiers.items():\n",
    "                    if key == \"translation\":\n",
    "                        continue\n",
    "                    #if key == \"ID\":\n",
    "                    #    continue\n",
    "                    if (attrs==None) or (key in attrs):\n",
    "                        if len(value) == 1:\n",
    "                            attributes_list.append((key, value[0]))\n",
    "                        else:\n",
    "                            attributes_list.append((key, \"; \".join(value)))\n",
    "\n",
    "                attributes_dict = OrderedDict( attributes_list )\n",
    "                feature_lists.append([rec.id, 'Genbank', feature.type, part.start+1, part.end, '.', strand_dict.get(part.strand, \".\"), \".\", attributes_dict])\n",
    "        \n",
    "    df=pd.DataFrame(feature_lists, columns=[\"seq_id\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('gene_id', 'pDONR201_2'),\n",
       "             ('gene_type', 'bacteria'),\n",
       "             ('complete', 'true'),\n",
       "             ('gc', '50'),\n",
       "             ('length', '306'),\n",
       "             ('source', 'GeneMark.hmm2'),\n",
       "             ('score', '20.71'),\n",
       "             ('phase', '0'),\n",
       "             ('name', 'pDONR201_2'),\n",
       "             ('cds_id', '1264_-1_959'),\n",
       "             ('domainator_pdonr_hmms', 'CcdB (CcdB protein, 1.1e-32, 103.1)'),\n",
       "             ('Color', '#FF0000')])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb_path=os.path.join(data_path, \"colored_genbank.gb\")\n",
    "recs=SeqIO.parse(gb_path, \"genbank\")\n",
    "rec=next(recs)\n",
    "df=seqRecord_to_df(rec, feature_types=[\"CDS\"])\n",
    "df.loc[1][\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pDONR201_1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2</td>\n",
       "      <td>106</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'gene_type': 'bacteria'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pDONR201_1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>959</td>\n",
       "      <td>1264</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'gene_type': 'bacteria'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pDONR201_1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1266</td>\n",
       "      <td>1391</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'gene_type': 'bacteria'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pDONR201_1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1606</td>\n",
       "      <td>2265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>{'gene_type': 'bacteria'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pDONR201_1</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2916</td>\n",
       "      <td>3677</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>{'gene_type': 'bacteria'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id   source type  start   end score strand phase  \\\n",
       "0  pDONR201_1  Genbank  CDS      2   106     .      +     .   \n",
       "1  pDONR201_1  Genbank  CDS    959  1264     .      -     .   \n",
       "2  pDONR201_1  Genbank  CDS   1266  1391     .      -     .   \n",
       "3  pDONR201_1  Genbank  CDS   1606  2265     .      -     .   \n",
       "4  pDONR201_1  Genbank  CDS   2916  3677     .      +     .   \n",
       "\n",
       "                  attributes  \n",
       "0  {'gene_type': 'bacteria'}  \n",
       "1  {'gene_type': 'bacteria'}  \n",
       "2  {'gene_type': 'bacteria'}  \n",
       "3  {'gene_type': 'bacteria'}  \n",
       "4  {'gene_type': 'bacteria'}  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=seqRecord_to_df(rec, feature_types=[\"CDS\"], attributes={\"CDS\":[\"gene_type\"]})\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "def parse_recs(recs, # iterator over Bio.SeqRecord.SeqRecord\n",
    "                   seq_id: Optional[str] = None, # sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name\n",
    "                   first = True, # if True then return only the annotations for the first sequence (or the first with seq_id)\n",
    "                   bounds: Optional[tuple] = None, # (left limit, right limit)\n",
    "                   feature_types: Optional[list] = None, # list of feature types to extract\n",
    "                   attributes: Optional[Dict[str, List]] = None, # a dictionary with feature types as keys and a list of attributes to extract as values \n",
    "               )->Tuple[List[Seq], List[pd.DataFrame]]:\n",
    "\n",
    "    # read genbank file(s)\n",
    "    feature_dfs = [] # list of dataframes, one for each seq record used if seq_id == \"all\"\n",
    "    seqs = [] # list of Seqs\n",
    "    for rec in recs:\n",
    "        if seq_id == rec.id or seq_id is None:\n",
    "            df = seqRecord_to_df(rec, feature_types=feature_types, attributes=attributes)\n",
    "            if bounds is not None:\n",
    "                df = df.loc[(df.end>bounds[0]) & (df.start<bounds[1])]\n",
    "            \n",
    "            df = set_positions(df)\n",
    "            feature_dfs.append(df)\n",
    "            seqs.append(rec.seq)\n",
    "            if first or seq_id is not None: # we only want one\n",
    "                break\n",
    "\n",
    "    if len(feature_dfs) == 0:\n",
    "        raise EmptyDataFrame(\"The annotation DataFrame is empty. Check that the feature_types and seq_id are correct, and that bounds (if specified) fall within the size of your genome.\")\n",
    "    return seqs, feature_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export    \n",
    "def parse_genbank(gb_path, # path to the genbank file\n",
    "                  seq_id: Optional[str] = None, # sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name\n",
    "                  first = True, # if True then return only the annotations for the first sequence (or the first with seq_id)\n",
    "                  bounds: Optional[tuple] = None, # (left limit, right limit)\n",
    "                  feature_types: Optional[list] = None, # list of feature types to extract\n",
    "                  attributes: Optional[Dict[str, List]] = None, # a dictionary with feature types as keys and a list of attributes to extract as values \n",
    "                  )->Tuple[List[Seq], List[pd.DataFrame]]:\n",
    "\n",
    "    with open(gb_path,\"r\") as f:\n",
    "        recs = parse_recs(SeqIO.parse(f, \"genbank\"), seq_id, first, bounds, feature_types, attributes)\n",
    "    return recs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_path=os.path.join(data_path, \"colored_genbank.gb\")\n",
    "recs=SeqIO.parse(gb_path, \"genbank\")\n",
    "rec=next(recs)\n",
    "testid=rec.id\n",
    "\n",
    "seqs, dfs=parse_genbank(gb_path, seq_id = testid,\n",
    "                 feature_types=[\"CDS\", \"Domainator\"]\n",
    "                )\n",
    "\n",
    "# dfs[0].head()\n",
    "assert len(dfs) == len(seqs)\n",
    "assert len(dfs) == 1\n",
    "assert dfs[0].loc[0, \"seq_id\"] == \"pDONR201_1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_path=os.path.join(data_path, \"colored_genbank.gb\")\n",
    "seqs, dfs=parse_genbank(gb_path,\n",
    "                 seq_id=None,\n",
    "                 first=False,\n",
    "                 feature_types=[\"CDS\", \"Domainator\"]\n",
    "                )\n",
    "\n",
    "assert len(dfs) == len(seqs)\n",
    "assert len(dfs) == 4\n",
    "assert dfs[0].loc[0, \"seq_id\"] == \"pDONR201_1\"\n",
    "assert dfs[1].loc[0, \"seq_id\"] == \"pDONR201_2\"\n",
    "assert dfs[2].loc[0, \"seq_id\"] == \"pDONR201_3\"\n",
    "assert dfs[3].loc[0, \"seq_id\"] == \"pDONR201_4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "def inspect_feature_types(file_path: str, \n",
    "                          frmt: str #gff or genbank\n",
    "                          ):\n",
    "    \"\"\"Outputs a table that recapitulates the feature types and attributes available in the file.\"\"\"\n",
    "    \n",
    "    if frmt == \"genbank\":\n",
    "        _, dfs=parse_genbank(file_path)\n",
    "    elif frmt == \"gff\":\n",
    "        dfs=parse_gff(file_path)\n",
    "\n",
    "    table_data=[]\n",
    "    for df in dfs:\n",
    "        for t in set(df.type):\n",
    "            row=[t]\n",
    "            attributes = df.loc[df.type==t, \"attributes\"].iloc[0]\n",
    "            for attr in attributes:\n",
    "                row.append(attr)\n",
    "                table_data.append(row)\n",
    "                row=[\"\"]\n",
    "\n",
    "\n",
    "    df_output = pd.DataFrame(table_data, columns=[\"feature_type\", \"attributes\"])\n",
    "    display(HTML(df_output.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_type</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>exon</td>\n",
       "      <td>ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Parent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>anticodon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gbkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>inference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>locus_tag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDS</td>\n",
       "      <td>ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Parent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Dbxref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gbkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>inference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>locus_tag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>protein_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>transl_table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tRNA</td>\n",
       "      <td>ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Parent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>anticodon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gbkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>inference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>locus_tag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>region</td>\n",
       "      <td>ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Dbxref</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>country</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gbkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>mol_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>nat-host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gene</td>\n",
       "      <td>ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gbkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gene_biotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>locus_tag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>old_locus_tag</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inspect_feature_types(gff_path, \"gff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_type</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CDS</td>\n",
       "      <td>gene_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gene_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>partial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>gc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>phase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>cds_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>source</td>\n",
       "      <td>invitrogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Domainator</td>\n",
       "      <td>program</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>evalue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>score</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>cds_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>Color</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#| hide\n",
    "gb_path=os.path.join(data_path, \"colored_genbank.gb\")\n",
    "inspect_feature_types(gb_path,\"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "def split_string(string, max_length=10):\n",
    "    if len(string) <= max_length:\n",
    "        return string\n",
    "    else:\n",
    "        split_index = max_length\n",
    "        while split_index > 0 and string[split_index] != ' ':\n",
    "            split_index -= 1\n",
    "        if split_index == 0:\n",
    "            split_index = max_length  # If no suitable breaking point found, split at max_length\n",
    "        return string[:split_index] + '\\n' + split_string(string[split_index:].lstrip(), max_length)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a very long string that needs to be split\n",
      "into multiple lines because it exceeds 50\n",
      "characters.\n"
     ]
    }
   ],
   "source": [
    "#| hide\n",
    "# Example usage\n",
    "long_string = \"This is a very long string that needs to be split into multiple lines because it exceeds 50 characters.\"\n",
    "\n",
    "split_result = split_string(long_string, max_length=50)\n",
    "print(split_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "def in_wsl() -> bool:\n",
    "    return 'microsoft-standard' in uname().release"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "in_wsl()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "def add_extension(filename,extension=\"svg\"):\n",
    "    base_name, ext = os.path.splitext(filename)\n",
    "    if ext.lower() != '.'+extension:\n",
    "        filename += '.'+extension\n",
    "    return filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('test.svg', 'test.png')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "add_extension(\"test.svg\"), add_extension(\"test\",extension=\"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "def sort_list_dict(d: dict, #a dictionnary for which all values are of type list\n",
    "                   ref_list=\"xs\", #key of the list to use as a reference for sorting\n",
    "                   func=lambda x: x, #A custom function can be supplied to customize the sort order. Default is the identity function.\n",
    "                  ):\n",
    "    ks=list(d.keys())\n",
    "    ref_list_ix=ks.index(ref_list)\n",
    "    # Sort all the lists in the dictionary based on the values of the reference list\n",
    "    sorted_lists = sorted(zip(*[d[k] for k in ks]), key= lambda x: func(x[ref_list_ix]))\n",
    "\n",
    "    # Convert the sorted tuples back into separate lists\n",
    "    unzipped_lists = zip(*sorted_lists)\n",
    "\n",
    "    # Create a new dictionary with the same keys as the original dictionary, but with the sorted lists as values\n",
    "    d = {k: list(t) for k, t in zip(ks, unzipped_lists)}\n",
    "    return d\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xs': [[3, 1], [2, 5]], 'a': [3, 1], 'b': ['d', 'c']}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "d={'xs':[[2,5],[3,1]],'a':[1,3],'b':[\"c\",\"d\"]}\n",
    "sort_list_dict(d, ref_list='xs', func= lambda x: x[1]) #sort according to the second element of each element of list 'xs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "from bokeh.plotting import show as bk_show\n",
    "from bokeh.layouts import column, row\n",
    "from bokeh.io import output_notebook, reset_output\n",
    "from bokeh.plotting import save as bk_save #Need to rename the bokeh show function so that there is no confusion with GenomeBrowser.show\n",
    "from bokeh.plotting import output_file as bk_output_file #Need to rename the bokeh show function so that there is no confusion with GenomeBrowser.show\n",
    "from bokeh.io import output_notebook, reset_output, export_png, export_svgs, export_svg\n",
    "from svgutils import compose\n",
    "import os\n",
    "import warnings\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium import webdriver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "def _save(elements, heights, width, fname:str, title:str=\"Genome Plot\"):\n",
    "    base_name, ext = os.path.splitext(fname)\n",
    "    ext = ext.lower()\n",
    "    if ext not in {\".svg\", \".png\"}:\n",
    "        raise ValueError(f\"filename must end in svg or png, not {ext}\")\n",
    "    \n",
    "    reset_output()\n",
    "    bk_output_file(filename=fname, title=title)\n",
    "    \n",
    "    layout = column(elements)\n",
    "\n",
    "    if in_wsl():\n",
    "            ## Setup chrome options\n",
    "            chrome_options = Options()\n",
    "            chrome_options.add_argument(\"--headless\") # Ensure GUI is off\n",
    "            chrome_options.add_argument(\"--no-sandbox\")\n",
    "            chrome_options.add_argument(\"--disable-3d-apis\")\n",
    "            chrome_options.add_argument(\"--disable-blink-features\")\n",
    "            \n",
    "\n",
    "            homedir = os.path.expanduser(\"~\")\n",
    "            try:\n",
    "                    # webdriver_service = Service(f\"{homedir}/chromedriver/stable/chromedriver\")\n",
    "                    # browser = webdriver.Chrome(service=webdriver_service, options=chrome_options)\n",
    "                    browser = webdriver.Chrome(options=chrome_options)\n",
    "            except:\n",
    "                    warnings.warn(\"\"\"If using WSL you can install chromedriver following these instructions:https://scottspence.com/posts/use-chrome-in-ubuntu-wsl\n",
    "                                  Also make sure the chromedriver-binary python package has the same major version number as your chrome install.\n",
    "                                  Check the chrome version using: google-chrome --version\n",
    "                                   Then use pip to force install of a web driver with a compatible version, for example:\n",
    "                                   pip install --force-reinstall -v \"chromedriver-binary==121.0.6167.184.0\"\n",
    "                                   \"\"\")\n",
    "                    browser=None\n",
    "\n",
    "            \n",
    "    else:\n",
    "            browser=None\n",
    "\n",
    "    if ext == \".svg\":\n",
    "        #export_svg(layout, filename=fname)\n",
    "        export_svgs(layout, filename=fname, webdriver=browser)\n",
    "        if len(heights)>1: # TODO: what is this?\n",
    "            total_height=sum(heights)\n",
    "            svgelements=[compose.SVG(fname)]\n",
    "            offset=heights[0]\n",
    "            for i, height in enumerate(heights[1:]):\n",
    "                svgelements.append(\n",
    "                    compose.SVG(f\"{base_name}_{i+1}.svg\").move(0,offset)\n",
    "                )\n",
    "                offset+=height\n",
    "                \n",
    "            compose.Figure(width+50, # +50 accounts for axis and labels\n",
    "                           total_height, \n",
    "                           *svgelements).save(f\"{base_name}_composite.svg\")\n",
    "\n",
    "    else:\n",
    "        export_png(layout, filename=fname, webdriver=browser)\n",
    "    \n",
    "    reset_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from genomenotebook.browser import GenomeBrowser\n",
    "from genomenotebook.plot import GenomePlot\n",
    "from genomenotebook.data import get_example_data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "data_path = get_example_data_dir()\n",
    "fasta_path = os.path.join(data_path, \"MG1655_U00096.fasta\")\n",
    "gff_path = os.path.join(data_path, \"MG1655_U00096.gff3\")\n",
    "\n",
    "g=GenomeBrowser(gff_path=gff_path, fasta_path=fasta_path, bounds=(0,5000), search=False)\n",
    "track = g.add_track(height=100)\n",
    "track.scatter(data=pd.DataFrame(dict(x=np.arange(0,5000,100),y=np.sin(np.arange(0,5000,100)))), y=\"y\", pos=\"x\")\n",
    "\n",
    "plot = GenomePlot(g)\n",
    "plot._collect_elements()\n",
    "\n",
    "_save(plot.elements, [g.height], g.width, fname=\"test_p.png\")\n",
    "\n",
    "file_path = os.path.join(os.getcwd(), \"test_p.png\")\n",
    "assert os.path.getsize(file_path) > 0, f\"File is empty: {file_path}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "os.remove(\"test_p.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "def _save_html(elements, fname:str, title:str):\n",
    "    reset_output()\n",
    "    bk_output_file(filename=fname, title=title, mode='inline')\n",
    "    bk_save(column(elements))\n",
    "    reset_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "#| export\n",
    "def _gb_show(elements):\n",
    "    reset_output()\n",
    "    output_notebook(hide_banner=True)\n",
    "    bk_show(column(elements))\n",
    "    reset_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
