<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="This contains useful functions">

<title>utils – genomenotebook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-0c9eb0277ab99f059f9e902403bd7a3d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="utils – genomenotebook">
<meta property="og:description" content="This contains useful functions">
<meta property="og:site_name" content="genomenotebook">
<meta name="twitter:title" content="utils – genomenotebook">
<meta name="twitter:description" content="This contains useful functions">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">genomenotebook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/dbikard/genomenotebook"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/browser.html">API</a></li><li class="breadcrumb-item"><a href="../API/utils.html">utils</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Examples</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../options.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Options</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../exporting plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exporting plots</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../stacking.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multi genbank and stacking</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">API</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/browser.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">browser</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/track.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">track</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/glyphs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">glyphs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/utils.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">utils</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#download_file" id="toc-download_file" class="nav-link active" data-scroll-target="#download_file">download_file</a></li>
  <li><a href="#is_gzipped_file" id="toc-is_gzipped_file" class="nav-link" data-scroll-target="#is_gzipped_file">is_gzipped_file</a></li>
  <li><a href="#default_open_gz" id="toc-default_open_gz" class="nav-link" data-scroll-target="#default_open_gz">default_open_gz</a></li>
  <li><a href="#extract_attribute" id="toc-extract_attribute" class="nav-link" data-scroll-target="#extract_attribute">extract_attribute</a></li>
  <li><a href="#extract_all_attributes" id="toc-extract_all_attributes" class="nav-link" data-scroll-target="#extract_all_attributes">extract_all_attributes</a></li>
  <li><a href="#extract_attributes" id="toc-extract_attributes" class="nav-link" data-scroll-target="#extract_attributes">extract_attributes</a></li>
  <li><a href="#get_attributes" id="toc-get_attributes" class="nav-link" data-scroll-target="#get_attributes">get_attributes</a></li>
  <li><a href="#attributes_to_columns" id="toc-attributes_to_columns" class="nav-link" data-scroll-target="#attributes_to_columns">attributes_to_columns</a></li>
  <li><a href="#set_positions" id="toc-set_positions" class="nav-link" data-scroll-target="#set_positions">set_positions</a></li>
  <li><a href="#parse_gff" id="toc-parse_gff" class="nav-link" data-scroll-target="#parse_gff">parse_gff</a></li>
  <li><a href="#available_feature_types" id="toc-available_feature_types" class="nav-link" data-scroll-target="#available_feature_types">available_feature_types</a></li>
  <li><a href="#available_attributes" id="toc-available_attributes" class="nav-link" data-scroll-target="#available_attributes">available_attributes</a></li>
  <li><a href="#parse_fasta" id="toc-parse_fasta" class="nav-link" data-scroll-target="#parse_fasta">parse_fasta</a></li>
  <li><a href="#regions_overlap" id="toc-regions_overlap" class="nav-link" data-scroll-target="#regions_overlap">regions_overlap</a></li>
  <li><a href="#add_z_order" id="toc-add_z_order" class="nav-link" data-scroll-target="#add_z_order">add_z_order</a></li>
  <li><a href="#get_cds_name" id="toc-get_cds_name" class="nav-link" data-scroll-target="#get_cds_name">get_cds_name</a></li>
  <li><a href="#get_cds_unique_name" id="toc-get_cds_unique_name" class="nav-link" data-scroll-target="#get_cds_unique_name">get_cds_unique_name</a></li>
  <li><a href="#seqrecord_to_df" id="toc-seqrecord_to_df" class="nav-link" data-scroll-target="#seqrecord_to_df">seqRecord_to_df</a></li>
  <li><a href="#parse_recs" id="toc-parse_recs" class="nav-link" data-scroll-target="#parse_recs">parse_recs</a></li>
  <li><a href="#parse_genbank" id="toc-parse_genbank" class="nav-link" data-scroll-target="#parse_genbank">parse_genbank</a></li>
  <li><a href="#inspect_feature_types" id="toc-inspect_feature_types" class="nav-link" data-scroll-target="#inspect_feature_types">inspect_feature_types</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/dbikard/genomenotebook/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/browser.html">API</a></li><li class="breadcrumb-item"><a href="../API/utils.html">utils</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">utils</h1>
</div>

<div>
  <div class="description">
    This contains useful functions
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<div id="cell-2" class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/javascript">
(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;
    // Clean up Bokeh references
    if (id != null && id in Bokeh.index) {
      Bokeh.index[id].model.document.clear();
      delete Bokeh.index[id];
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim();
            if (id in Bokeh.index) {
              Bokeh.index[id].model.document.clear();
              delete Bokeh.index[id];
            }
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded() {
    const el = document.getElementById(null);
    if (el != null) {
      el.textContent = "BokehJS is loading...";
    }
    if (root.Bokeh !== undefined) {
      if (el != null) {
        el.textContent = "BokehJS " + root.Bokeh.version + " successfully loaded.";
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(display_loaded, 100)
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.2.2.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.2.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.2.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.2.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.2.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
          for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }
} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById(null)).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));
</script>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L34" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="download_file" class="level3">
<h3 class="anchored" data-anchor-id="download_file">download_file</h3>
<blockquote class="blockquote">
<pre><code> download_file (url, save_path)</code></pre>
</blockquote>
<p><em>Checks if a file with the same name is already in the save_path. If not download it.</em></p>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L43" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="is_gzipped_file" class="level3">
<h3 class="anchored" data-anchor-id="is_gzipped_file">is_gzipped_file</h3>
<blockquote class="blockquote">
<pre><code> is_gzipped_file (file_path)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L53" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="default_open_gz" class="level3">
<h3 class="anchored" data-anchor-id="default_open_gz">default_open_gz</h3>
<blockquote class="blockquote">
<pre><code> default_open_gz (gff_path)</code></pre>
</blockquote>
<p><em>If file is gzipped then opens it with <code>gzip.open</code>, otherwise opens it with <code>open</code></em></p>
<div id="cell-6" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example usage</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>file_url <span class="op">=</span> <span class="st">'https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gff.gz'</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>human_genome_gff <span class="op">=</span> <span class="st">'GRCh38_latest_genomic.gff.gz'</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>download_file(file_url, human_genome_gff)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>is_gzipped_file(human_genome_gff)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>File already exists: GRCh38_latest_genomic.gff.gz</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>True</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L61" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="extract_attribute" class="level3">
<h3 class="anchored" data-anchor-id="extract_attribute">extract_attribute</h3>
<blockquote class="blockquote">
<pre><code> extract_attribute (input_str:str, attr_name:str)</code></pre>
</blockquote>
<p><em>Extracts the attribute called attr_name from the GFF attributes string</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>input_str</td>
<td>str</td>
<td>attribute string to parse</td>
</tr>
<tr class="even">
<td>attr_name</td>
<td>str</td>
<td>name of the attribute to extract</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-8" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>input_str <span class="op">=</span> <span class="st">'ID=cds-ATV02827.1;Parent=gene-SaO11_00001;Dbxref=NCBI_GP:ATV02827.1;Name=ATV02827.1;gbkey=CDS;gene=dnaA;locus_tag=SaO11_00001;product=Chromosomal replication initiator protein DnaA;protein_id=ATV02827.1;transl_table=11'</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>extract_attribute(input_str,<span class="st">"gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'dnaA'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L74" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="extract_all_attributes" class="level3">
<h3 class="anchored" data-anchor-id="extract_all_attributes">extract_all_attributes</h3>
<blockquote class="blockquote">
<pre><code> extract_all_attributes (input_str:str)</code></pre>
</blockquote>
<p><em>Extracts all attributes from the GFF attributes column</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>input_str</td>
<td>str</td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>OrderedDict</strong></td>
<td><strong>TODO: why is this not limited by the attributes subset provided to GenomeBrowser?</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L84" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="extract_attributes" class="level3">
<h3 class="anchored" data-anchor-id="extract_attributes">extract_attributes</h3>
<blockquote class="blockquote">
<pre><code> extract_attributes (input_str:str, attributes:Optional[List[str]]=None)</code></pre>
</blockquote>
<p><em>Extracts attributes from the GFF attributes column</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>input_str</td>
<td>str</td>
<td></td>
<td>the attribute string of a GFF fome</td>
</tr>
<tr class="even">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td>an optional list of attribute names to extract. If None all attributes are extracted.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>OrderedDict</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L97" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_attributes" class="level3">
<h3 class="anchored" data-anchor-id="get_attributes">get_attributes</h3>
<blockquote class="blockquote">
<pre><code> get_attributes (df:pandas.core.frame.DataFrame,
                 attributes:Optional[Dict[str,List]]=None)</code></pre>
</blockquote>
<p><em>Iterates over each row of the df and extracts the attributes specified in the attributes dictionary for each feature type</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>df</td>
<td>DataFrame</td>
<td></td>
<td>a features DataFrame with at least a “type” column and an “attributes_str” column</td>
</tr>
<tr class="even">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td>a dictionary with feature types as keys and a list of attributes to extract as values</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>List</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L115" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="attributes_to_columns" class="level3">
<h3 class="anchored" data-anchor-id="attributes_to_columns">attributes_to_columns</h3>
<blockquote class="blockquote">
<pre><code> attributes_to_columns (features:pandas.core.frame.DataFrame)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L129" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="set_positions" class="level3">
<h3 class="anchored" data-anchor-id="set_positions">set_positions</h3>
<blockquote class="blockquote">
<pre><code> set_positions (annotation:pandas.core.frame.DataFrame)</code></pre>
</blockquote>
<p><em>Sets left and right as the position of the feature on the sequence, left is always lower than right. start and end represent the begining and end of the feature where start can be greater than end depending on the feature strand.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>annotation</td>
<td>DataFrame</td>
<td>an annotation DataFrame extracted from a gff file</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L155" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_gff" class="level3">
<h3 class="anchored" data-anchor-id="parse_gff">parse_gff</h3>
<blockquote class="blockquote">
<pre><code> parse_gff (gff_path:str, seq_id:Optional[str]=None, first:bool=True,
            bounds:Optional[tuple]=None,
            feature_types:Optional[list]=None,
            attributes:Optional[Dict[str,List]]=None)</code></pre>
</blockquote>
<p><em>Parses a GFF3 file and returns a list of Pandas DataFrames with the data for a specific contig. If seq_id is None then only the first contig is parsed. If feature_types is None then all feature types are extracted.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gff_path</td>
<td>str</td>
<td></td>
<td>path to the gff file</td>
</tr>
<tr class="even">
<td>seq_id</td>
<td>Optional</td>
<td>None</td>
<td>sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name</td>
</tr>
<tr class="odd">
<td>first</td>
<td>bool</td>
<td>True</td>
<td>if True then return only the annotations for the first sequence (or the first with seq_id)</td>
</tr>
<tr class="even">
<td>bounds</td>
<td>Optional</td>
<td>None</td>
<td>(left limit, right limit)</td>
</tr>
<tr class="odd">
<td>feature_types</td>
<td>Optional</td>
<td>None</td>
<td>list of feature types to extract</td>
</tr>
<tr class="even">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td>a dictionary with feature types as keys and a list of attributes to extract as values</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>List</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> genomenotebook.data <span class="im">import</span> get_example_data_dir</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-17" class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>data_path <span class="op">=</span> get_example_data_dir()</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>gff_path <span class="op">=</span> os.path.join(data_path, <span class="st">"jmh43.gff"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>df<span class="op">=</span>parse_gff(gff_path, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>             bounds<span class="op">=</span>(<span class="dv">10000</span>,<span class="dv">50000</span>))[<span class="dv">0</span>]</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seq_id</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">left</th>
<th data-quarto-table-cell-role="th">right</th>
<th data-quarto-table-cell-role="th">middle</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>NZ_JAGURL010000100.1</td>
<td>RefSeq</td>
<td>region</td>
<td>1</td>
<td>16949</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'NZ_JAGURL010000100.1:1..16949', 'Dbxre...</td>
<td>1</td>
<td>16949</td>
<td>8475.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>NZ_JAGURL010000100.1</td>
<td>RefSeq</td>
<td>gene</td>
<td>10059</td>
<td>8311</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>{'ID': 'gene-KFX61_RS20985', 'Name': 'KFX61_RS...</td>
<td>8311</td>
<td>10059</td>
<td>9185.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>NZ_JAGURL010000100.1</td>
<td>Protein Homology</td>
<td>CDS</td>
<td>10059</td>
<td>8311</td>
<td>.</td>
<td>-</td>
<td>0</td>
<td>{'ID': 'cds-WP_225638104.1', 'Parent': 'gene-K...</td>
<td>8311</td>
<td>10059</td>
<td>9185.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>NZ_JAGURL010000100.1</td>
<td>RefSeq</td>
<td>gene</td>
<td>10541</td>
<td>10092</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>{'ID': 'gene-KFX61_RS20990', 'Name': 'KFX61_RS...</td>
<td>10092</td>
<td>10541</td>
<td>10316.5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>NZ_JAGURL010000100.1</td>
<td>Protein Homology</td>
<td>CDS</td>
<td>10541</td>
<td>10092</td>
<td>.</td>
<td>-</td>
<td>0</td>
<td>{'ID': 'cds-WP_016268480.1', 'Parent': 'gene-K...</td>
<td>10092</td>
<td>10541</td>
<td>10316.5</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-18" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>df<span class="op">=</span>parse_gff(gff_path, </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>             seq_id<span class="op">=</span><span class="st">"NZ_JAGURL010000013.1"</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>             bounds<span class="op">=</span>(<span class="dv">10000</span>,<span class="dv">50000</span>))[<span class="dv">0</span>]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seq_id</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">left</th>
<th data-quarto-table-cell-role="th">right</th>
<th data-quarto-table-cell-role="th">middle</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>NZ_JAGURL010000013.1</td>
<td>RefSeq</td>
<td>region</td>
<td>1</td>
<td>98838</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'NZ_JAGURL010000013.1:1..98838', 'Dbxre...</td>
<td>1</td>
<td>98838</td>
<td>49419.5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>NZ_JAGURL010000013.1</td>
<td>RefSeq</td>
<td>gene</td>
<td>10195</td>
<td>11175</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'gene-KFX61_RS05880', 'Name': 'KFX61_RS...</td>
<td>10195</td>
<td>11175</td>
<td>10685.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>NZ_JAGURL010000013.1</td>
<td>Protein Homology</td>
<td>CDS</td>
<td>10195</td>
<td>11175</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>{'ID': 'cds-WP_048697200.1', 'Parent': 'gene-K...</td>
<td>10195</td>
<td>11175</td>
<td>10685.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>NZ_JAGURL010000013.1</td>
<td>RefSeq</td>
<td>gene</td>
<td>11234</td>
<td>11824</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'gene-KFX61_RS05885', 'Name': 'KFX61_RS...</td>
<td>11234</td>
<td>11824</td>
<td>11529.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>NZ_JAGURL010000013.1</td>
<td>Protein Homology</td>
<td>CDS</td>
<td>11234</td>
<td>11824</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>{'ID': 'cds-WP_048697198.1', 'Parent': 'gene-K...</td>
<td>11234</td>
<td>11824</td>
<td>11529.0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-19" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>genome_path <span class="op">=</span> os.path.join(data_path, <span class="st">"MG1655_U00096.fasta"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>gff_path <span class="op">=</span> os.path.join(data_path, <span class="st">"MG1655_U00096.gff3"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>parse_gff(gff_path)[<span class="dv">0</span>].head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seq_id</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">left</th>
<th data-quarto-table-cell-role="th">right</th>
<th data-quarto-table-cell-role="th">middle</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>U00096.3</td>
<td>Genbank</td>
<td>region</td>
<td>1</td>
<td>4641652</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'U00096.3:1..4641652', 'Dbxref': 'taxon...</td>
<td>1</td>
<td>4641652</td>
<td>2320826.5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>U00096.3</td>
<td>Genbank</td>
<td>gene</td>
<td>190</td>
<td>255</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'gene-b0001', 'Dbxref': 'ASAP:ABE-00000...</td>
<td>190</td>
<td>255</td>
<td>222.5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>U00096.3</td>
<td>Genbank</td>
<td>CDS</td>
<td>190</td>
<td>255</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>{'ID': 'cds-AAC73112.1', 'Parent': 'gene-b0001...</td>
<td>190</td>
<td>255</td>
<td>222.5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>U00096.3</td>
<td>Genbank</td>
<td>gene</td>
<td>337</td>
<td>2799</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'ID': 'gene-b0002', 'Dbxref': 'ASAP:ABE-00000...</td>
<td>337</td>
<td>2799</td>
<td>1568.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>U00096.3</td>
<td>Genbank</td>
<td>CDS</td>
<td>337</td>
<td>2799</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>{'ID': 'cds-AAC73113.1', 'Parent': 'gene-b0002...</td>
<td>337</td>
<td>2799</td>
<td>1568.0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L224" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="available_feature_types" class="level3">
<h3 class="anchored" data-anchor-id="available_feature_types">available_feature_types</h3>
<blockquote class="blockquote">
<pre><code> available_feature_types (gff_path)</code></pre>
</blockquote>
<div id="cell-21" class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>data_path <span class="op">=</span> get_example_data_dir()</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>gff_path <span class="op">=</span> os.path.join(data_path, <span class="st">"MG1655_U00096.gff3"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>available_feature_types(gff_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'CDS',
 'exon',
 'gene',
 'mobile_genetic_element',
 'ncRNA',
 'origin_of_replication',
 'pseudogene',
 'rRNA',
 'recombination_feature',
 'region',
 'repeat_region',
 'sequence_feature',
 'tRNA'}</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L235" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="available_attributes" class="level3">
<h3 class="anchored" data-anchor-id="available_attributes">available_attributes</h3>
<blockquote class="blockquote">
<pre><code> available_attributes (gff_path)</code></pre>
</blockquote>
<div id="cell-23" class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>available_attributes(gff_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Index(['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase',
       'attributes', 'left', 'right', 'middle'],
      dtype='object')</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L240" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_fasta" class="level3">
<h3 class="anchored" data-anchor-id="parse_fasta">parse_fasta</h3>
<blockquote class="blockquote">
<pre><code> parse_fasta (genome_path, seq_id)</code></pre>
</blockquote>
<p><em>Retrieves the Biopython SeqRecord object that matches the seq_id in a fasta file</em></p>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L257" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="regions_overlap" class="level3">
<h3 class="anchored" data-anchor-id="regions_overlap">regions_overlap</h3>
<blockquote class="blockquote">
<pre><code> regions_overlap (region1, region2, min_overlap_fraction=0.0)</code></pre>
</blockquote>
<p><em>regions are tuples of start and stop coordinates returns true if a fraction of region2 &gt;= min_overlap_fraction overlaps with region1 coordinates within regions must be sorted low to high</em></p>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L313" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_z_order" class="level3">
<h3 class="anchored" data-anchor-id="add_z_order">add_z_order</h3>
<blockquote class="blockquote">
<pre><code> add_z_order (features, prescedence=['source', 'CDS', 'repeat_region',
              'ncRNA', 'rRNA', 'tRNA', 'exon'])</code></pre>
</blockquote>
<p><em>features is a dataframe of features prescedence is a list of feature types in order of prescedence, e.g.&nbsp;[“CDS”, “repeat_region”, “ncRNA”, “rRNA”, “tRNA”] will put “CDS” features closer to the bottom of the plot than “repeat_region” features. returns features with a z_order column added</em></p>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L363" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_cds_name" class="level3">
<h3 class="anchored" data-anchor-id="get_cds_name">get_cds_name</h3>
<blockquote class="blockquote">
<pre><code> get_cds_name (feature)</code></pre>
</blockquote>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>feature</td>
<td>(contig_id, feature):</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L352" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_cds_unique_name" class="level3">
<h3 class="anchored" data-anchor-id="get_cds_unique_name">get_cds_unique_name</h3>
<blockquote class="blockquote">
<pre><code> get_cds_unique_name (feature)</code></pre>
</blockquote>
<p><em>If the feature already has a cds_id, then keep it, otherwise generate one based on the position on the contig.</em></p>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L378" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="seqrecord_to_df" class="level3">
<h3 class="anchored" data-anchor-id="seqrecord_to_df">seqRecord_to_df</h3>
<blockquote class="blockquote">
<pre><code> seqRecord_to_df (rec:&lt;module'Bio.SeqRecord'from'/opt/hostedtoolcache/Pyth
                  on/3.10.17/x64/lib/python3.10/site-
                  packages/Bio/SeqRecord.py'&gt;,
                  feature_types:Optional[List[str]]=None,
                  attributes:Optional[Dict[str,List]]=None)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>rec</td>
<td>Bio.SeqRecord</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>feature_types</td>
<td>Optional</td>
<td>None</td>
<td>if None then get all features, otherwise only those with type in FeatureTypes.</td>
</tr>
<tr class="odd">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>if None, then get all attributes of all feature types. If dict, then only get attributes of feature types keys. If value is None, get all</strong></td>
</tr>
</tbody>
</table>
<div id="cell-30" class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>gb_path<span class="op">=</span>os.path.join(data_path, <span class="st">"colored_genbank.gb"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>recs<span class="op">=</span>SeqIO.parse(gb_path, <span class="st">"genbank"</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>rec<span class="op">=</span><span class="bu">next</span>(recs)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>df<span class="op">=</span>seqRecord_to_df(rec, feature_types<span class="op">=</span>[<span class="st">"CDS"</span>])</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>df.loc[<span class="dv">1</span>][<span class="st">"attributes"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>OrderedDict([('gene_id', 'pDONR201_2'),
             ('gene_type', 'bacteria'),
             ('complete', 'true'),
             ('gc', '50'),
             ('length', '306'),
             ('source', 'GeneMark.hmm2'),
             ('score', '20.71'),
             ('phase', '0'),
             ('name', 'pDONR201_2'),
             ('cds_id', '1264_-1_959'),
             ('domainator_pdonr_hmms', 'CcdB (CcdB protein, 1.1e-32, 103.1)'),
             ('Color', '#FF0000')])</code></pre>
</div>
</div>
<div id="cell-31" class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>df<span class="op">=</span>seqRecord_to_df(rec, feature_types<span class="op">=</span>[<span class="st">"CDS"</span>], attributes<span class="op">=</span>{<span class="st">"CDS"</span>:[<span class="st">"gene_type"</span>]})</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seq_id</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>pDONR201_1</td>
<td>Genbank</td>
<td>CDS</td>
<td>2</td>
<td>106</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'gene_type': 'bacteria'}</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pDONR201_1</td>
<td>Genbank</td>
<td>CDS</td>
<td>959</td>
<td>1264</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>{'gene_type': 'bacteria'}</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>pDONR201_1</td>
<td>Genbank</td>
<td>CDS</td>
<td>1266</td>
<td>1391</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>{'gene_type': 'bacteria'}</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>pDONR201_1</td>
<td>Genbank</td>
<td>CDS</td>
<td>1606</td>
<td>2265</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>{'gene_type': 'bacteria'}</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>pDONR201_1</td>
<td>Genbank</td>
<td>CDS</td>
<td>2916</td>
<td>3677</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>{'gene_type': 'bacteria'}</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L411" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_recs" class="level3">
<h3 class="anchored" data-anchor-id="parse_recs">parse_recs</h3>
<blockquote class="blockquote">
<pre><code> parse_recs (recs, seq_id:Optional[str]=None, first=True,
             bounds:Optional[tuple]=None,
             feature_types:Optional[list]=None,
             attributes:Optional[Dict[str,List]]=None)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>recs</td>
<td></td>
<td></td>
<td>iterator over Bio.SeqRecord.SeqRecord</td>
</tr>
<tr class="even">
<td>seq_id</td>
<td>Optional</td>
<td>None</td>
<td>sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name</td>
</tr>
<tr class="odd">
<td>first</td>
<td>bool</td>
<td>True</td>
<td>if True then return only the annotations for the first sequence (or the first with seq_id)</td>
</tr>
<tr class="even">
<td>bounds</td>
<td>Optional</td>
<td>None</td>
<td>(left limit, right limit)</td>
</tr>
<tr class="odd">
<td>feature_types</td>
<td>Optional</td>
<td>None</td>
<td>list of feature types to extract</td>
</tr>
<tr class="even">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td>a dictionary with feature types as keys and a list of attributes to extract as values</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>Tuple</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L439" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_genbank" class="level3">
<h3 class="anchored" data-anchor-id="parse_genbank">parse_genbank</h3>
<blockquote class="blockquote">
<pre><code> parse_genbank (gb_path, seq_id:Optional[str]=None, first=True,
                bounds:Optional[tuple]=None,
                feature_types:Optional[list]=None,
                attributes:Optional[Dict[str,List]]=None)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gb_path</td>
<td></td>
<td></td>
<td>path to the genbank file</td>
</tr>
<tr class="even">
<td>seq_id</td>
<td>Optional</td>
<td>None</td>
<td>sequence id (first column of the gff), if not None, then return only the annotations for the seq_id with this name</td>
</tr>
<tr class="odd">
<td>first</td>
<td>bool</td>
<td>True</td>
<td>if True then return only the annotations for the first sequence (or the first with seq_id)</td>
</tr>
<tr class="even">
<td>bounds</td>
<td>Optional</td>
<td>None</td>
<td>(left limit, right limit)</td>
</tr>
<tr class="odd">
<td>feature_types</td>
<td>Optional</td>
<td>None</td>
<td>list of feature types to extract</td>
</tr>
<tr class="even">
<td>attributes</td>
<td>Optional</td>
<td>None</td>
<td>a dictionary with feature types as keys and a list of attributes to extract as values</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>Tuple</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-34" class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>gb_path<span class="op">=</span>os.path.join(data_path, <span class="st">"colored_genbank.gb"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>recs<span class="op">=</span>SeqIO.parse(gb_path, <span class="st">"genbank"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>rec<span class="op">=</span><span class="bu">next</span>(recs)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>testid<span class="op">=</span>rec.<span class="bu">id</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>seqs, dfs<span class="op">=</span>parse_genbank(gb_path, seq_id <span class="op">=</span> testid,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>                 feature_types<span class="op">=</span>[<span class="st">"CDS"</span>, <span class="st">"Domainator"</span>]</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co"># dfs[0].head()</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(dfs) <span class="op">==</span> <span class="bu">len</span>(seqs)</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(dfs) <span class="op">==</span> <span class="dv">1</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> dfs[<span class="dv">0</span>].loc[<span class="dv">0</span>, <span class="st">"seq_id"</span>] <span class="op">==</span> <span class="st">"pDONR201_1"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-35" class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>gb_path<span class="op">=</span>os.path.join(data_path, <span class="st">"colored_genbank.gb"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>seqs, dfs<span class="op">=</span>parse_genbank(gb_path,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                 seq_id<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                 first<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                 feature_types<span class="op">=</span>[<span class="st">"CDS"</span>, <span class="st">"Domainator"</span>]</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(dfs) <span class="op">==</span> <span class="bu">len</span>(seqs)</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(dfs) <span class="op">==</span> <span class="dv">4</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> dfs[<span class="dv">0</span>].loc[<span class="dv">0</span>, <span class="st">"seq_id"</span>] <span class="op">==</span> <span class="st">"pDONR201_1"</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> dfs[<span class="dv">1</span>].loc[<span class="dv">0</span>, <span class="st">"seq_id"</span>] <span class="op">==</span> <span class="st">"pDONR201_2"</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> dfs[<span class="dv">2</span>].loc[<span class="dv">0</span>, <span class="st">"seq_id"</span>] <span class="op">==</span> <span class="st">"pDONR201_3"</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> dfs[<span class="dv">3</span>].loc[<span class="dv">0</span>, <span class="st">"seq_id"</span>] <span class="op">==</span> <span class="st">"pDONR201_4"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/dbikard/genomenotebook/blob/main/genomenotebook/utils.py#L453" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="inspect_feature_types" class="level3">
<h3 class="anchored" data-anchor-id="inspect_feature_types">inspect_feature_types</h3>
<blockquote class="blockquote">
<pre><code> inspect_feature_types (file_path:str, frmt:str)</code></pre>
</blockquote>
<p><em>Outputs a table that recapitulates the feature types and attributes available in the file.</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>file_path</td>
<td>str</td>
<td></td>
</tr>
<tr class="even">
<td>frmt</td>
<td>str</td>
<td>gff or genbank</td>
</tr>
</tbody>
</table>
<div id="cell-37" class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>inspect_feature_types(gff_path, <span class="st">"gff"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">feature_type</th>
<th data-quarto-table-cell-role="th">attributes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>exon</td>
<td>ID</td>
</tr>
<tr class="even">
<td></td>
<td>Parent</td>
</tr>
<tr class="odd">
<td></td>
<td>anticodon</td>
</tr>
<tr class="even">
<td></td>
<td>gbkey</td>
</tr>
<tr class="odd">
<td></td>
<td>inference</td>
</tr>
<tr class="even">
<td></td>
<td>locus_tag</td>
</tr>
<tr class="odd">
<td></td>
<td>product</td>
</tr>
<tr class="even">
<td>CDS</td>
<td>ID</td>
</tr>
<tr class="odd">
<td></td>
<td>Parent</td>
</tr>
<tr class="even">
<td></td>
<td>Dbxref</td>
</tr>
<tr class="odd">
<td></td>
<td>Name</td>
</tr>
<tr class="even">
<td></td>
<td>gbkey</td>
</tr>
<tr class="odd">
<td></td>
<td>inference</td>
</tr>
<tr class="even">
<td></td>
<td>locus_tag</td>
</tr>
<tr class="odd">
<td></td>
<td>product</td>
</tr>
<tr class="even">
<td></td>
<td>protein_id</td>
</tr>
<tr class="odd">
<td></td>
<td>transl_table</td>
</tr>
<tr class="even">
<td>tRNA</td>
<td>ID</td>
</tr>
<tr class="odd">
<td></td>
<td>Parent</td>
</tr>
<tr class="even">
<td></td>
<td>anticodon</td>
</tr>
<tr class="odd">
<td></td>
<td>gbkey</td>
</tr>
<tr class="even">
<td></td>
<td>inference</td>
</tr>
<tr class="odd">
<td></td>
<td>locus_tag</td>
</tr>
<tr class="even">
<td></td>
<td>product</td>
</tr>
<tr class="odd">
<td>region</td>
<td>ID</td>
</tr>
<tr class="even">
<td></td>
<td>Dbxref</td>
</tr>
<tr class="odd">
<td></td>
<td>country</td>
</tr>
<tr class="even">
<td></td>
<td>gbkey</td>
</tr>
<tr class="odd">
<td></td>
<td>genome</td>
</tr>
<tr class="even">
<td></td>
<td>mol_type</td>
</tr>
<tr class="odd">
<td></td>
<td>nat-host</td>
</tr>
<tr class="even">
<td></td>
<td>strain</td>
</tr>
<tr class="odd">
<td>gene</td>
<td>ID</td>
</tr>
<tr class="even">
<td></td>
<td>Name</td>
</tr>
<tr class="odd">
<td></td>
<td>gbkey</td>
</tr>
<tr class="even">
<td></td>
<td>gene_biotype</td>
</tr>
<tr class="odd">
<td></td>
<td>locus_tag</td>
</tr>
<tr class="even">
<td></td>
<td>old_locus_tag</td>
</tr>
</tbody>
</table>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/dbikard\.github\.io\/genomenotebook");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dbikard/genomenotebook/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>